{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "anonymous-tourism",
   "metadata": {},
   "source": [
    "# Machine Learning for disease diagnosis using omics data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "swiss-addition",
   "metadata": {},
   "source": [
    "This file and raw data in github repo here: https://github.com/parkyed/sepsis_ml_omics_msc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cheap-garbage",
   "metadata": {},
   "source": [
    "# Code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "hydraulic-washer",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import stats\n",
    "from collections import OrderedDict\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_curve, auc, roc_auc_score\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay, plot_confusion_matrix\n",
    "from sklearn.model_selection import KFold, RepeatedKFold\n",
    "import warnings\n",
    "from joblib import dump, load\n",
    "from pickle import dump, load\n",
    "warnings.filterwarnings(action=\"ignore\", message=\"^internal gelsd\")\n",
    "warnings.filterwarnings(action=\"ignore\", message=\"Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.\")\n",
    "#np.random.seed(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "behavioral-civilization",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/Ed/Documents/GitHub/sepsis_ml_omics_msc'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lesser-connectivity",
   "metadata": {},
   "source": [
    "### Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "handed-ebony",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of columns is:  93\n",
      "The number of row is:  48804\n"
     ]
    }
   ],
   "source": [
    "# import and check data\n",
    "\n",
    "raw_data = pd.read_csv('genomic_data.csv')\n",
    "print('The number of columns is:  '+str(len(raw_data.columns)))\n",
    "print('The number of row is:  '+str(len(raw_data)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "speaking-mistake",
   "metadata": {},
   "source": [
    "### Checking for Duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "rising-bleeding",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# duplicate rows based on NuID:   0\n",
      "# duplicate rows based on Search_Key:   4671\n",
      "# duplicate rows based on ILMN_Gene:   11000\n",
      "# duplicate rows based on RefSeq_ID:   5742\n",
      "# duplicate rows based on Entrez_Gene_ID:   10897\n",
      "# duplicate rows based on Probe_Id:   1\n"
     ]
    }
   ],
   "source": [
    "# Counting the number of duplicate rows using various columns, NaN values excluded before looking for duplicates.\n",
    "\n",
    "dup_column = ['NuID', 'Search_Key', 'ILMN_Gene', 'RefSeq_ID', 'Entrez_Gene_ID', 'Probe_Id']\n",
    "for column in dup_column:\n",
    "    data = raw_data.dropna(subset=[column])\n",
    "    dup_list = []\n",
    "    for index, value in data.duplicated(subset=[column]).items():\n",
    "        if value == True:\n",
    "            dup_list.append(index)\n",
    "    print(f\"# duplicate rows based on {column}:   \"+str(len(dup_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "administrative-quality",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NuID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>Entrez_Gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Inf_149</th>\n",
       "      <th>Inf_152</th>\n",
       "      <th>Inf_157</th>\n",
       "      <th>Inf_159a</th>\n",
       "      <th>Inf_162a</th>\n",
       "      <th>Inf_164</th>\n",
       "      <th>Inf_191</th>\n",
       "      <th>Inf_198</th>\n",
       "      <th>Inf_203</th>\n",
       "      <th>Fold change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48794</th>\n",
       "      <td>ZZQvowA1A0rdICPUkk</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_24212</td>\n",
       "      <td>ILMN_24212</td>\n",
       "      <td>CASP8AP2</td>\n",
       "      <td>NM_012115.2</td>\n",
       "      <td>NM_012115.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9994.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0826</td>\n",
       "      <td>7.1113</td>\n",
       "      <td>7.0518</td>\n",
       "      <td>7.0041</td>\n",
       "      <td>7.0160</td>\n",
       "      <td>6.9972</td>\n",
       "      <td>7.1369</td>\n",
       "      <td>7.0338</td>\n",
       "      <td>7.0494</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48795</th>\n",
       "      <td>ZzTi0SkIUngIr1kHu0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_137958</td>\n",
       "      <td>ILMN_37701</td>\n",
       "      <td>C20ORF7</td>\n",
       "      <td>NM_024120.3</td>\n",
       "      <td>NM_024120.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79133.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.8497</td>\n",
       "      <td>7.7296</td>\n",
       "      <td>7.5873</td>\n",
       "      <td>7.4077</td>\n",
       "      <td>7.4651</td>\n",
       "      <td>7.4088</td>\n",
       "      <td>7.6038</td>\n",
       "      <td>7.5458</td>\n",
       "      <td>7.5341</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48796</th>\n",
       "      <td>ZzTla5Up6pZQklKEU4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_27978</td>\n",
       "      <td>ILMN_27978</td>\n",
       "      <td>TMEM151A</td>\n",
       "      <td>NM_153266.2</td>\n",
       "      <td>NM_153266.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256472.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0177</td>\n",
       "      <td>6.9288</td>\n",
       "      <td>7.0070</td>\n",
       "      <td>6.9190</td>\n",
       "      <td>6.9754</td>\n",
       "      <td>6.9154</td>\n",
       "      <td>7.0130</td>\n",
       "      <td>7.0644</td>\n",
       "      <td>6.9691</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48797</th>\n",
       "      <td>ZZTZrt5dKGKbmWEeks</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_31400</td>\n",
       "      <td>ILMN_31400</td>\n",
       "      <td>LOC645781</td>\n",
       "      <td>XM_933141.2</td>\n",
       "      <td>XM_933141.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>645781.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1248</td>\n",
       "      <td>7.0515</td>\n",
       "      <td>7.0518</td>\n",
       "      <td>7.1918</td>\n",
       "      <td>7.0641</td>\n",
       "      <td>7.0524</td>\n",
       "      <td>6.9758</td>\n",
       "      <td>7.0188</td>\n",
       "      <td>7.0494</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48798</th>\n",
       "      <td>ZztzUERXUExGBETFFU</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_14674</td>\n",
       "      <td>ILMN_14674</td>\n",
       "      <td>NPY6R</td>\n",
       "      <td>NR_002713.1</td>\n",
       "      <td>NR_002713.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0010</td>\n",
       "      <td>7.0157</td>\n",
       "      <td>6.9841</td>\n",
       "      <td>6.9733</td>\n",
       "      <td>7.0813</td>\n",
       "      <td>7.0127</td>\n",
       "      <td>6.9264</td>\n",
       "      <td>7.0644</td>\n",
       "      <td>6.9361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48799</th>\n",
       "      <td>ZzVW43OPaE6GO1WJqs</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_139314</td>\n",
       "      <td>ILMN_139314</td>\n",
       "      <td>SEZ6L2</td>\n",
       "      <td>XM_939114.1</td>\n",
       "      <td>XM_939114.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26470.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1036</td>\n",
       "      <td>7.0593</td>\n",
       "      <td>7.0298</td>\n",
       "      <td>7.1770</td>\n",
       "      <td>7.1478</td>\n",
       "      <td>7.0254</td>\n",
       "      <td>7.0245</td>\n",
       "      <td>6.9904</td>\n",
       "      <td>6.9919</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48800</th>\n",
       "      <td>ZZXgKFKE5q5rp53.p0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_42654</td>\n",
       "      <td>ILMN_42654</td>\n",
       "      <td>LOC339047</td>\n",
       "      <td>XM_932631.1</td>\n",
       "      <td>XM_932631.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>339047.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0509</td>\n",
       "      <td>7.0305</td>\n",
       "      <td>6.9952</td>\n",
       "      <td>7.0171</td>\n",
       "      <td>7.0033</td>\n",
       "      <td>7.0127</td>\n",
       "      <td>7.1234</td>\n",
       "      <td>7.0338</td>\n",
       "      <td>7.0380</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48801</th>\n",
       "      <td>ZzXXie6icAfD693B6w</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_104503</td>\n",
       "      <td>ILMN_104503</td>\n",
       "      <td>HS.539415</td>\n",
       "      <td>Hs.539415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.539415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0423</td>\n",
       "      <td>7.0072</td>\n",
       "      <td>7.0298</td>\n",
       "      <td>7.0441</td>\n",
       "      <td>7.0160</td>\n",
       "      <td>7.0254</td>\n",
       "      <td>7.0245</td>\n",
       "      <td>7.1131</td>\n",
       "      <td>7.0494</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48802</th>\n",
       "      <td>ZzZ5OziA4Anaop_fcU</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_14103</td>\n",
       "      <td>ILMN_165051</td>\n",
       "      <td>MAGI1</td>\n",
       "      <td>NM_015520.1</td>\n",
       "      <td>NM_015520.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9223.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0423</td>\n",
       "      <td>7.0593</td>\n",
       "      <td>7.1600</td>\n",
       "      <td>7.0308</td>\n",
       "      <td>7.0294</td>\n",
       "      <td>7.0672</td>\n",
       "      <td>7.0606</td>\n",
       "      <td>7.0644</td>\n",
       "      <td>6.9804</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48803</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_8135</td>\n",
       "      <td>ILMN_8135</td>\n",
       "      <td>TUBB2A</td>\n",
       "      <td>NM_001069.2</td>\n",
       "      <td>NM_001069.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7280.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     NuID       Species   Source   Search_Key   Transcript  \\\n",
       "48794  ZZQvowA1A0rdICPUkk  Homo sapiens   RefSeq   ILMN_24212   ILMN_24212   \n",
       "48795  ZzTi0SkIUngIr1kHu0  Homo sapiens   RefSeq  ILMN_137958   ILMN_37701   \n",
       "48796  ZzTla5Up6pZQklKEU4  Homo sapiens   RefSeq   ILMN_27978   ILMN_27978   \n",
       "48797  ZZTZrt5dKGKbmWEeks  Homo sapiens   RefSeq   ILMN_31400   ILMN_31400   \n",
       "48798  ZztzUERXUExGBETFFU  Homo sapiens   RefSeq   ILMN_14674   ILMN_14674   \n",
       "48799  ZzVW43OPaE6GO1WJqs  Homo sapiens   RefSeq  ILMN_139314  ILMN_139314   \n",
       "48800  ZZXgKFKE5q5rp53.p0  Homo sapiens   RefSeq   ILMN_42654   ILMN_42654   \n",
       "48801  ZzXXie6icAfD693B6w  Homo sapiens  Unigene  ILMN_104503  ILMN_104503   \n",
       "48802  ZzZ5OziA4Anaop_fcU  Homo sapiens   RefSeq   ILMN_14103  ILMN_165051   \n",
       "48803                 NaN  Homo sapiens   RefSeq    ILMN_8135    ILMN_8135   \n",
       "\n",
       "       ILMN_Gene Source_Reference_ID    RefSeq_ID Unigene_ID  Entrez_Gene_ID  \\\n",
       "48794   CASP8AP2         NM_012115.2  NM_012115.2        NaN          9994.0   \n",
       "48795    C20ORF7         NM_024120.3  NM_024120.3        NaN         79133.0   \n",
       "48796   TMEM151A         NM_153266.2  NM_153266.2        NaN        256472.0   \n",
       "48797  LOC645781         XM_933141.2  XM_933141.2        NaN        645781.0   \n",
       "48798      NPY6R         NR_002713.1  NR_002713.1        NaN             NaN   \n",
       "48799     SEZ6L2         XM_939114.1  XM_939114.1        NaN         26470.0   \n",
       "48800  LOC339047         XM_932631.1  XM_932631.1        NaN        339047.0   \n",
       "48801  HS.539415           Hs.539415          NaN  Hs.539415             NaN   \n",
       "48802      MAGI1         NM_015520.1  NM_015520.1        NaN          9223.0   \n",
       "48803     TUBB2A         NM_001069.2  NM_001069.2        NaN          7280.0   \n",
       "\n",
       "       ...  Inf_149 Inf_152 Inf_157 Inf_159a Inf_162a  Inf_164 Inf_191  \\\n",
       "48794  ...   7.0826  7.1113  7.0518   7.0041   7.0160   6.9972  7.1369   \n",
       "48795  ...   7.8497  7.7296  7.5873   7.4077   7.4651   7.4088  7.6038   \n",
       "48796  ...   7.0177  6.9288  7.0070   6.9190   6.9754   6.9154  7.0130   \n",
       "48797  ...   7.1248  7.0515  7.0518   7.1918   7.0641   7.0524  6.9758   \n",
       "48798  ...   7.0010  7.0157  6.9841   6.9733   7.0813   7.0127  6.9264   \n",
       "48799  ...   7.1036  7.0593  7.0298   7.1770   7.1478   7.0254  7.0245   \n",
       "48800  ...   7.0509  7.0305  6.9952   7.0171   7.0033   7.0127  7.1234   \n",
       "48801  ...   7.0423  7.0072  7.0298   7.0441   7.0160   7.0254  7.0245   \n",
       "48802  ...   7.0423  7.0593  7.1600   7.0308   7.0294   7.0672  7.0606   \n",
       "48803  ...      NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "\n",
       "       Inf_198 Inf_203 Fold change  \n",
       "48794   7.0338  7.0494         NaN  \n",
       "48795   7.5458  7.5341         NaN  \n",
       "48796   7.0644  6.9691         NaN  \n",
       "48797   7.0188  7.0494         NaN  \n",
       "48798   7.0644  6.9361         NaN  \n",
       "48799   6.9904  6.9919         NaN  \n",
       "48800   7.0338  7.0380         NaN  \n",
       "48801   7.1131  7.0494         NaN  \n",
       "48802   7.0644  6.9804         NaN  \n",
       "48803      NaN     NaN         NaN  \n",
       "\n",
       "[10 rows x 93 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "suited-integer",
   "metadata": {},
   "source": [
    "### Drop duplicate rows on Probe_Id and transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "fuzzy-supply",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of features in dataset:   48802\n",
      "Number of genes in the gene name database:    48802\n",
      "Number of examples in dataset:  63\n",
      "Number of Missing values in dataset:   0\n",
      "The number of features with all zeros is:   0\n",
      "Check indices of df and gene_df match. sum of matching index:   48802\n"
     ]
    }
   ],
   "source": [
    "# reformat data, drop duplicates, transpose, index on probe_id, drop columns with all NaN, drop Fold change\n",
    "\n",
    "# drop duplicate rows based on Probe_Id, and fold change column\n",
    "df = raw_data.drop_duplicates(subset=['Probe_Id'], keep='first')\n",
    "df = df.drop(['Fold change'], axis=1)\n",
    "\n",
    "# split into gene code dataframe, and data set based on unique probes\n",
    "gene_df = df.iloc[:, 4:15]\n",
    "df = df.iloc[:, np.r_[14, 29:92]]\n",
    "\n",
    "# transpose data set and index on probe ID, drop all columns that are all NaN values\n",
    "df = df.transpose()\n",
    "df = df.rename(columns=df.iloc[0]).drop(df.index[0])\n",
    "\n",
    "# identify any columns that are all NaN values, and drop from both data df and gene_df\n",
    "nan_columns = df.columns[df.isna().any()].tolist()\n",
    "df = df.drop(nan_columns, axis=1)\n",
    "gene_df = gene_df[gene_df['Probe_Id'] != nan_columns[0]]\n",
    "print('Number of features in dataset:   '+str(len(df.columns)))\n",
    "print('Number of genes in the gene name database:    '+str(len(gene_df.index)))\n",
    "print('Number of examples in dataset:  '+str(len(df)))\n",
    "print('Number of Missing values in dataset:   '+str(df.isnull().sum().sum()))\n",
    "print('The number of features with all zeros is:   '+str(sum((df == 0).all(axis=0))))\n",
    "print('Check indices of df and gene_df match. sum of matching index:   '+str(sum(df.columns == gene_df['Probe_Id'])))\n",
    "\n",
    "# create copy of the data without patient Inf_075\n",
    "df_no75 = df.drop(['Inf075'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "digital-helen",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create labels for both data sets\n",
    "\n",
    "labels = []\n",
    "for item in df.index:\n",
    "    if 'Con' in item:\n",
    "        labels.append(0)\n",
    "    else:\n",
    "        labels.append(1)\n",
    "labels = np.asarray(labels)\n",
    "\n",
    "labels_no75 = []\n",
    "for item in df.index:\n",
    "    if item == 'Inf075':\n",
    "        continue\n",
    "    if 'Con' in item:\n",
    "        labels_no75.append(0)\n",
    "    else:\n",
    "        labels_no75.append(1)\n",
    "labels_no75 = np.asarray(labels_no75)\n",
    "\n",
    "# check correct\n",
    "#list(zip(df.index, labels))\n",
    "#list(zip(df_no75.index, labels_no75))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "crucial-density",
   "metadata": {},
   "source": [
    "### Standardisation (z-scoring)\n",
    "\n",
    "Data standardised (i.e. z-scored) for the following reasons:\n",
    "- L1 regularisation penalties in logistic regressions assumes data centred at zero and on the same scale\n",
    "- Distance based ML models such as SVM require and assume standardised data, otherwise variables on larger scales disproportionally impact the model\n",
    "- Am using the output coeffiecients from logistic regression as a crude measure of feature importance, and so in order to compare coefficients as a measure of relative importance, variables must be standardised\n",
    "\n",
    "Reference: https://towardsdatascience.com/normalization-vs-standardization-quantitative-analysis-a91e8a79cebf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "adolescent-denial",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63, 48802)\n",
      "(62, 48802)\n",
      "(63, 200)\n"
     ]
    }
   ],
   "source": [
    "# standardise by imputing NaN values and using standard scale - z-scoring\n",
    "# note - given imputer not necessary, could move standard scaler into the pipeline for each model\n",
    "\n",
    "def standardise(examples):\n",
    "    scaler = StandardScaler()\n",
    "    examples_scaled = scaler.fit_transform(examples)\n",
    "    return examples_scaled\n",
    "\n",
    "X_df = standardise(df)\n",
    "X_df = pd.DataFrame(X_df, columns=df.columns, index=df.index)\n",
    "print(X_df.shape)\n",
    "\n",
    "# dataset excluding patient 75\n",
    "X_df_no75 = standardise(df_no75)\n",
    "X_df_no75 = pd.DataFrame(X_df_no75, columns=df_no75.columns, index=df_no75.index)\n",
    "print(X_df_no75.shape)\n",
    "\n",
    "# Reduced dataset for code testing\n",
    "X_df_red = X_df.iloc[:, 0:200]\n",
    "print(X_df_red.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "approximate-supply",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retriving saved results:\n",
    "import pickle\n",
    "\n",
    "with open('neonatal_preprocessed.pkl', 'wb') as f:\n",
    "    pickle.dump([gene_df, X_df, X_df_no75, X_df_red, labels, labels_no75], f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "subject-forward",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'numpy.ndarray' object has no attribute 'to_csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-94-dd46dd8f61d5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mX_df_no75\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'neonatal_data_processed.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mlabels_no75\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'labels_no75.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'to_csv'"
     ]
    }
   ],
   "source": [
    "X_df_no75.to_csv('neonatal_data_processed.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "requested-preview",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ILMN_2227757</th>\n",
       "      <th>ILMN_1683690</th>\n",
       "      <th>ILMN_1710146</th>\n",
       "      <th>ILMN_1822171</th>\n",
       "      <th>ILMN_1748473</th>\n",
       "      <th>ILMN_1822843</th>\n",
       "      <th>ILMN_1710170</th>\n",
       "      <th>ILMN_1704497</th>\n",
       "      <th>ILMN_1847638</th>\n",
       "      <th>ILMN_1772260</th>\n",
       "      <th>...</th>\n",
       "      <th>ILMN_1712859</th>\n",
       "      <th>ILMN_1747129</th>\n",
       "      <th>ILMN_1813344</th>\n",
       "      <th>ILMN_2292576</th>\n",
       "      <th>ILMN_1793445</th>\n",
       "      <th>ILMN_2065273</th>\n",
       "      <th>ILMN_1800873</th>\n",
       "      <th>ILMN_1666642</th>\n",
       "      <th>ILMN_1908209</th>\n",
       "      <th>ILMN_1670343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Con_001</th>\n",
       "      <td>0.479050</td>\n",
       "      <td>-0.138235</td>\n",
       "      <td>0.190287</td>\n",
       "      <td>0.417313</td>\n",
       "      <td>0.912011</td>\n",
       "      <td>0.315686</td>\n",
       "      <td>-0.739573</td>\n",
       "      <td>-0.127496</td>\n",
       "      <td>-0.931788</td>\n",
       "      <td>1.009489</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.553490</td>\n",
       "      <td>0.216461</td>\n",
       "      <td>-0.283713</td>\n",
       "      <td>0.543436</td>\n",
       "      <td>-0.679029</td>\n",
       "      <td>-0.269263</td>\n",
       "      <td>1.260044</td>\n",
       "      <td>3.326382</td>\n",
       "      <td>4.054684</td>\n",
       "      <td>-1.112666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_017</th>\n",
       "      <td>-0.796642</td>\n",
       "      <td>0.171890</td>\n",
       "      <td>0.377130</td>\n",
       "      <td>1.360378</td>\n",
       "      <td>0.480996</td>\n",
       "      <td>1.673514</td>\n",
       "      <td>-0.649962</td>\n",
       "      <td>1.807527</td>\n",
       "      <td>-1.601138</td>\n",
       "      <td>1.506726</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.739596</td>\n",
       "      <td>-0.698123</td>\n",
       "      <td>1.807757</td>\n",
       "      <td>-0.504326</td>\n",
       "      <td>-0.941835</td>\n",
       "      <td>0.970388</td>\n",
       "      <td>-0.260336</td>\n",
       "      <td>-1.397208</td>\n",
       "      <td>-0.561287</td>\n",
       "      <td>-0.247056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_021</th>\n",
       "      <td>0.842998</td>\n",
       "      <td>-0.018567</td>\n",
       "      <td>0.593805</td>\n",
       "      <td>0.619009</td>\n",
       "      <td>0.456604</td>\n",
       "      <td>0.620670</td>\n",
       "      <td>0.151053</td>\n",
       "      <td>-0.561857</td>\n",
       "      <td>-0.544413</td>\n",
       "      <td>-0.213574</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.364277</td>\n",
       "      <td>-0.708119</td>\n",
       "      <td>0.645692</td>\n",
       "      <td>-0.081346</td>\n",
       "      <td>1.121258</td>\n",
       "      <td>0.433380</td>\n",
       "      <td>-0.036937</td>\n",
       "      <td>-0.868595</td>\n",
       "      <td>0.949872</td>\n",
       "      <td>-0.995543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_022b</th>\n",
       "      <td>-0.245093</td>\n",
       "      <td>-0.158461</td>\n",
       "      <td>-0.271325</td>\n",
       "      <td>-0.442166</td>\n",
       "      <td>-0.266965</td>\n",
       "      <td>0.917543</td>\n",
       "      <td>-0.106808</td>\n",
       "      <td>1.068957</td>\n",
       "      <td>-0.573158</td>\n",
       "      <td>-0.625975</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053212</td>\n",
       "      <td>1.810735</td>\n",
       "      <td>0.668380</td>\n",
       "      <td>-1.900804</td>\n",
       "      <td>-0.050579</td>\n",
       "      <td>-0.947497</td>\n",
       "      <td>0.687527</td>\n",
       "      <td>0.666900</td>\n",
       "      <td>-0.418795</td>\n",
       "      <td>-0.199475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_028</th>\n",
       "      <td>-0.751618</td>\n",
       "      <td>-1.341659</td>\n",
       "      <td>-0.445607</td>\n",
       "      <td>0.631728</td>\n",
       "      <td>0.619020</td>\n",
       "      <td>-1.074588</td>\n",
       "      <td>-2.045337</td>\n",
       "      <td>1.627197</td>\n",
       "      <td>1.559456</td>\n",
       "      <td>2.419899</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.640495</td>\n",
       "      <td>-0.876375</td>\n",
       "      <td>1.898924</td>\n",
       "      <td>0.160818</td>\n",
       "      <td>0.081459</td>\n",
       "      <td>-1.059083</td>\n",
       "      <td>0.088782</td>\n",
       "      <td>0.813555</td>\n",
       "      <td>3.000998</td>\n",
       "      <td>-0.457511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inf_162a</th>\n",
       "      <td>0.873014</td>\n",
       "      <td>0.697756</td>\n",
       "      <td>0.749246</td>\n",
       "      <td>-0.965485</td>\n",
       "      <td>-0.391260</td>\n",
       "      <td>-0.724180</td>\n",
       "      <td>0.167512</td>\n",
       "      <td>-0.696712</td>\n",
       "      <td>0.367218</td>\n",
       "      <td>0.396779</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.397311</td>\n",
       "      <td>-0.704787</td>\n",
       "      <td>-1.140515</td>\n",
       "      <td>-0.733574</td>\n",
       "      <td>0.004014</td>\n",
       "      <td>0.708858</td>\n",
       "      <td>0.554573</td>\n",
       "      <td>-0.805117</td>\n",
       "      <td>-0.568786</td>\n",
       "      <td>-0.259866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inf_164</th>\n",
       "      <td>-1.430736</td>\n",
       "      <td>-1.547286</td>\n",
       "      <td>-0.729797</td>\n",
       "      <td>-1.178083</td>\n",
       "      <td>0.579584</td>\n",
       "      <td>2.053122</td>\n",
       "      <td>-0.735915</td>\n",
       "      <td>0.531102</td>\n",
       "      <td>0.022276</td>\n",
       "      <td>0.441554</td>\n",
       "      <td>...</td>\n",
       "      <td>1.287407</td>\n",
       "      <td>-1.017977</td>\n",
       "      <td>-1.372763</td>\n",
       "      <td>-1.702229</td>\n",
       "      <td>-0.144529</td>\n",
       "      <td>-0.487205</td>\n",
       "      <td>-0.552473</td>\n",
       "      <td>-0.702240</td>\n",
       "      <td>-0.392547</td>\n",
       "      <td>0.431890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inf_191</th>\n",
       "      <td>-0.162548</td>\n",
       "      <td>0.861246</td>\n",
       "      <td>-0.161417</td>\n",
       "      <td>0.871583</td>\n",
       "      <td>-2.291614</td>\n",
       "      <td>-0.338083</td>\n",
       "      <td>-1.215061</td>\n",
       "      <td>0.802381</td>\n",
       "      <td>1.729189</td>\n",
       "      <td>-0.402100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.915602</td>\n",
       "      <td>1.309296</td>\n",
       "      <td>-0.568351</td>\n",
       "      <td>-0.126550</td>\n",
       "      <td>-1.117040</td>\n",
       "      <td>-1.991873</td>\n",
       "      <td>-0.560613</td>\n",
       "      <td>0.509301</td>\n",
       "      <td>-0.409421</td>\n",
       "      <td>0.311107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inf_198</th>\n",
       "      <td>-0.505859</td>\n",
       "      <td>0.522468</td>\n",
       "      <td>1.414972</td>\n",
       "      <td>0.867949</td>\n",
       "      <td>-0.322321</td>\n",
       "      <td>-0.961030</td>\n",
       "      <td>0.867943</td>\n",
       "      <td>-0.914677</td>\n",
       "      <td>0.161896</td>\n",
       "      <td>0.068037</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118282</td>\n",
       "      <td>-0.408255</td>\n",
       "      <td>-0.807612</td>\n",
       "      <td>0.703265</td>\n",
       "      <td>-0.571113</td>\n",
       "      <td>0.414201</td>\n",
       "      <td>-0.869030</td>\n",
       "      <td>-0.471314</td>\n",
       "      <td>1.251729</td>\n",
       "      <td>0.380649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inf_203</th>\n",
       "      <td>1.722850</td>\n",
       "      <td>1.525320</td>\n",
       "      <td>1.452655</td>\n",
       "      <td>0.898839</td>\n",
       "      <td>-0.194228</td>\n",
       "      <td>-1.558020</td>\n",
       "      <td>-0.086691</td>\n",
       "      <td>-1.349038</td>\n",
       "      <td>3.945301</td>\n",
       "      <td>0.769118</td>\n",
       "      <td>...</td>\n",
       "      <td>3.772664</td>\n",
       "      <td>-0.148373</td>\n",
       "      <td>-0.855877</td>\n",
       "      <td>-0.835283</td>\n",
       "      <td>-0.182617</td>\n",
       "      <td>-1.822750</td>\n",
       "      <td>-0.855463</td>\n",
       "      <td>-0.425348</td>\n",
       "      <td>0.057426</td>\n",
       "      <td>-1.156587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 48802 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ILMN_2227757  ILMN_1683690  ILMN_1710146  ILMN_1822171  \\\n",
       "Con_001       0.479050     -0.138235      0.190287      0.417313   \n",
       "Con_017      -0.796642      0.171890      0.377130      1.360378   \n",
       "Con_021       0.842998     -0.018567      0.593805      0.619009   \n",
       "Con_022b     -0.245093     -0.158461     -0.271325     -0.442166   \n",
       "Con_028      -0.751618     -1.341659     -0.445607      0.631728   \n",
       "...                ...           ...           ...           ...   \n",
       "Inf_162a      0.873014      0.697756      0.749246     -0.965485   \n",
       "Inf_164      -1.430736     -1.547286     -0.729797     -1.178083   \n",
       "Inf_191      -0.162548      0.861246     -0.161417      0.871583   \n",
       "Inf_198      -0.505859      0.522468      1.414972      0.867949   \n",
       "Inf_203       1.722850      1.525320      1.452655      0.898839   \n",
       "\n",
       "          ILMN_1748473  ILMN_1822843  ILMN_1710170  ILMN_1704497  \\\n",
       "Con_001       0.912011      0.315686     -0.739573     -0.127496   \n",
       "Con_017       0.480996      1.673514     -0.649962      1.807527   \n",
       "Con_021       0.456604      0.620670      0.151053     -0.561857   \n",
       "Con_022b     -0.266965      0.917543     -0.106808      1.068957   \n",
       "Con_028       0.619020     -1.074588     -2.045337      1.627197   \n",
       "...                ...           ...           ...           ...   \n",
       "Inf_162a     -0.391260     -0.724180      0.167512     -0.696712   \n",
       "Inf_164       0.579584      2.053122     -0.735915      0.531102   \n",
       "Inf_191      -2.291614     -0.338083     -1.215061      0.802381   \n",
       "Inf_198      -0.322321     -0.961030      0.867943     -0.914677   \n",
       "Inf_203      -0.194228     -1.558020     -0.086691     -1.349038   \n",
       "\n",
       "          ILMN_1847638  ILMN_1772260  ...  ILMN_1712859  ILMN_1747129  \\\n",
       "Con_001      -0.931788      1.009489  ...     -1.553490      0.216461   \n",
       "Con_017      -1.601138      1.506726  ...     -0.739596     -0.698123   \n",
       "Con_021      -0.544413     -0.213574  ...     -0.364277     -0.708119   \n",
       "Con_022b     -0.573158     -0.625975  ...      0.053212      1.810735   \n",
       "Con_028       1.559456      2.419899  ...     -0.640495     -0.876375   \n",
       "...                ...           ...  ...           ...           ...   \n",
       "Inf_162a      0.367218      0.396779  ...     -0.397311     -0.704787   \n",
       "Inf_164       0.022276      0.441554  ...      1.287407     -1.017977   \n",
       "Inf_191       1.729189     -0.402100  ...      0.915602      1.309296   \n",
       "Inf_198       0.161896      0.068037  ...     -0.118282     -0.408255   \n",
       "Inf_203       3.945301      0.769118  ...      3.772664     -0.148373   \n",
       "\n",
       "          ILMN_1813344  ILMN_2292576  ILMN_1793445  ILMN_2065273  \\\n",
       "Con_001      -0.283713      0.543436     -0.679029     -0.269263   \n",
       "Con_017       1.807757     -0.504326     -0.941835      0.970388   \n",
       "Con_021       0.645692     -0.081346      1.121258      0.433380   \n",
       "Con_022b      0.668380     -1.900804     -0.050579     -0.947497   \n",
       "Con_028       1.898924      0.160818      0.081459     -1.059083   \n",
       "...                ...           ...           ...           ...   \n",
       "Inf_162a     -1.140515     -0.733574      0.004014      0.708858   \n",
       "Inf_164      -1.372763     -1.702229     -0.144529     -0.487205   \n",
       "Inf_191      -0.568351     -0.126550     -1.117040     -1.991873   \n",
       "Inf_198      -0.807612      0.703265     -0.571113      0.414201   \n",
       "Inf_203      -0.855877     -0.835283     -0.182617     -1.822750   \n",
       "\n",
       "          ILMN_1800873  ILMN_1666642  ILMN_1908209  ILMN_1670343  \n",
       "Con_001       1.260044      3.326382      4.054684     -1.112666  \n",
       "Con_017      -0.260336     -1.397208     -0.561287     -0.247056  \n",
       "Con_021      -0.036937     -0.868595      0.949872     -0.995543  \n",
       "Con_022b      0.687527      0.666900     -0.418795     -0.199475  \n",
       "Con_028       0.088782      0.813555      3.000998     -0.457511  \n",
       "...                ...           ...           ...           ...  \n",
       "Inf_162a      0.554573     -0.805117     -0.568786     -0.259866  \n",
       "Inf_164      -0.552473     -0.702240     -0.392547      0.431890  \n",
       "Inf_191      -0.560613      0.509301     -0.409421      0.311107  \n",
       "Inf_198      -0.869030     -0.471314      1.251729      0.380649  \n",
       "Inf_203      -0.855463     -0.425348      0.057426     -1.156587  \n",
       "\n",
       "[62 rows x 48802 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('neonatal_data_processed.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "grave-private",
   "metadata": {},
   "outputs": [],
   "source": [
    "savetxt('labels_no75.csv', labels_no75, delimiter=',')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "swedish-chemistry",
   "metadata": {},
   "source": [
    "### De-duplicating based on ensembl_gene_id\n",
    "\n",
    "NOTE: NOT YET DONE, MAY NEED TO GET THE IDS USING R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "yellow-appraisal",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NuID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>Entrez_Gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Inf_149</th>\n",
       "      <th>Inf_152</th>\n",
       "      <th>Inf_157</th>\n",
       "      <th>Inf_159a</th>\n",
       "      <th>Inf_162a</th>\n",
       "      <th>Inf_164</th>\n",
       "      <th>Inf_191</th>\n",
       "      <th>Inf_198</th>\n",
       "      <th>Inf_203</th>\n",
       "      <th>Fold change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25103</th>\n",
       "      <td>Ku8QhfS0n_hIOABXuE</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_5579</td>\n",
       "      <td>ILMN_183371</td>\n",
       "      <td>7A5</td>\n",
       "      <td>NM_182762.2</td>\n",
       "      <td>NM_182762.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>346389.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0267</td>\n",
       "      <td>6.9027</td>\n",
       "      <td>6.9288</td>\n",
       "      <td>6.9657</td>\n",
       "      <td>7.0408</td>\n",
       "      <td>6.9972</td>\n",
       "      <td>6.9992</td>\n",
       "      <td>6.9750</td>\n",
       "      <td>7.0849</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29049</th>\n",
       "      <td>NlLnpDrjjo8PUQKCJI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_19974</td>\n",
       "      <td>ILMN_19974</td>\n",
       "      <td>A26B1</td>\n",
       "      <td>NM_207355.2</td>\n",
       "      <td>NM_207355.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>339010.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0348</td>\n",
       "      <td>6.9117</td>\n",
       "      <td>7.0070</td>\n",
       "      <td>7.0041</td>\n",
       "      <td>7.1023</td>\n",
       "      <td>7.0822</td>\n",
       "      <td>7.0845</td>\n",
       "      <td>7.0491</td>\n",
       "      <td>7.0978</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42813</th>\n",
       "      <td>Wje_nHHp916e8Akjro</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_31305</td>\n",
       "      <td>ILMN_307810</td>\n",
       "      <td>A26C1B</td>\n",
       "      <td>NM_001099771.1</td>\n",
       "      <td>NM_001099771.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>728378.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0592</td>\n",
       "      <td>7.0157</td>\n",
       "      <td>7.0641</td>\n",
       "      <td>6.9264</td>\n",
       "      <td>7.0590</td>\n",
       "      <td>7.1122</td>\n",
       "      <td>7.0724</td>\n",
       "      <td>7.1493</td>\n",
       "      <td>7.0153</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47175</th>\n",
       "      <td>ZfoDn4IUQp4Auf56LU</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_8023</td>\n",
       "      <td>ILMN_8023</td>\n",
       "      <td>A2M</td>\n",
       "      <td>NM_000014.4</td>\n",
       "      <td>NM_000014.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9565</td>\n",
       "      <td>7.0456</td>\n",
       "      <td>7.0070</td>\n",
       "      <td>7.0510</td>\n",
       "      <td>6.9608</td>\n",
       "      <td>7.0524</td>\n",
       "      <td>6.9637</td>\n",
       "      <td>6.9195</td>\n",
       "      <td>7.1239</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7232</th>\n",
       "      <td>97DZJV_gpQpKINl0Ec</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_17375</td>\n",
       "      <td>ILMN_17375</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_144670.2</td>\n",
       "      <td>NM_144670.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>144568.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0741</td>\n",
       "      <td>7.1184</td>\n",
       "      <td>7.0750</td>\n",
       "      <td>7.0510</td>\n",
       "      <td>6.9889</td>\n",
       "      <td>6.9835</td>\n",
       "      <td>7.0724</td>\n",
       "      <td>7.0491</td>\n",
       "      <td>7.0849</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31612</th>\n",
       "      <td>opLIq5393IogH5SiEI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_24644</td>\n",
       "      <td>ILMN_24644</td>\n",
       "      <td>ZUFSP</td>\n",
       "      <td>NM_145062.1</td>\n",
       "      <td>NM_145062.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>221302.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.1171</td>\n",
       "      <td>7.7784</td>\n",
       "      <td>7.7894</td>\n",
       "      <td>8.1040</td>\n",
       "      <td>8.1978</td>\n",
       "      <td>7.8296</td>\n",
       "      <td>7.7331</td>\n",
       "      <td>7.8357</td>\n",
       "      <td>7.8272</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>05KVyf6Big_oj1KLSo</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_22056</td>\n",
       "      <td>ILMN_22056</td>\n",
       "      <td>ZW10</td>\n",
       "      <td>NM_004724.2</td>\n",
       "      <td>NM_004724.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9183.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.9831</td>\n",
       "      <td>8.1483</td>\n",
       "      <td>7.9581</td>\n",
       "      <td>7.2240</td>\n",
       "      <td>7.5738</td>\n",
       "      <td>7.8844</td>\n",
       "      <td>7.7331</td>\n",
       "      <td>7.6729</td>\n",
       "      <td>7.5341</td>\n",
       "      <td>-0.3903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15920</th>\n",
       "      <td>EVdUtxU1JfTj09e7Qo</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_4827</td>\n",
       "      <td>ILMN_4827</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>NM_007157.3</td>\n",
       "      <td>NM_007157.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>158586.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1934</td>\n",
       "      <td>7.1242</td>\n",
       "      <td>7.2487</td>\n",
       "      <td>7.0580</td>\n",
       "      <td>7.0641</td>\n",
       "      <td>7.1840</td>\n",
       "      <td>7.1504</td>\n",
       "      <td>7.3298</td>\n",
       "      <td>7.1376</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24628</th>\n",
       "      <td>KoL5XxK6hf_7w3v91U</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_16610</td>\n",
       "      <td>ILMN_16610</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>NM_024646.1</td>\n",
       "      <td>NM_024646.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79699.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0346</td>\n",
       "      <td>8.6077</td>\n",
       "      <td>8.3911</td>\n",
       "      <td>8.4578</td>\n",
       "      <td>8.6710</td>\n",
       "      <td>8.2491</td>\n",
       "      <td>8.1199</td>\n",
       "      <td>8.1083</td>\n",
       "      <td>8.6916</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28926</th>\n",
       "      <td>NKKYJWSJAroBG5cZ7A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_11865</td>\n",
       "      <td>ILMN_11865</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>NM_015113.3</td>\n",
       "      <td>NM_015113.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23140.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.8563</td>\n",
       "      <td>8.4258</td>\n",
       "      <td>8.2854</td>\n",
       "      <td>7.7498</td>\n",
       "      <td>8.3944</td>\n",
       "      <td>8.3913</td>\n",
       "      <td>8.4547</td>\n",
       "      <td>8.1570</td>\n",
       "      <td>8.3454</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30804 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     NuID       Species  Source  Search_Key   Transcript  \\\n",
       "25103  Ku8QhfS0n_hIOABXuE  Homo sapiens  RefSeq   ILMN_5579  ILMN_183371   \n",
       "29049  NlLnpDrjjo8PUQKCJI  Homo sapiens  RefSeq  ILMN_19974   ILMN_19974   \n",
       "42813  Wje_nHHp916e8Akjro  Homo sapiens  RefSeq  ILMN_31305  ILMN_307810   \n",
       "47175  ZfoDn4IUQp4Auf56LU  Homo sapiens  RefSeq   ILMN_8023    ILMN_8023   \n",
       "7232   97DZJV_gpQpKINl0Ec  Homo sapiens  RefSeq  ILMN_17375   ILMN_17375   \n",
       "...                   ...           ...     ...         ...          ...   \n",
       "31612  opLIq5393IogH5SiEI  Homo sapiens  RefSeq  ILMN_24644   ILMN_24644   \n",
       "252    05KVyf6Big_oj1KLSo  Homo sapiens  RefSeq  ILMN_22056   ILMN_22056   \n",
       "15920  EVdUtxU1JfTj09e7Qo  Homo sapiens  RefSeq   ILMN_4827    ILMN_4827   \n",
       "24628  KoL5XxK6hf_7w3v91U  Homo sapiens  RefSeq  ILMN_16610   ILMN_16610   \n",
       "28926  NKKYJWSJAroBG5cZ7A  Homo sapiens  RefSeq  ILMN_11865   ILMN_11865   \n",
       "\n",
       "      ILMN_Gene Source_Reference_ID       RefSeq_ID Unigene_ID  \\\n",
       "25103       7A5         NM_182762.2     NM_182762.2        NaN   \n",
       "29049     A26B1         NM_207355.2     NM_207355.2        NaN   \n",
       "42813    A26C1B      NM_001099771.1  NM_001099771.1        NaN   \n",
       "47175       A2M         NM_000014.4     NM_000014.4        NaN   \n",
       "7232      A2ML1         NM_144670.2     NM_144670.2        NaN   \n",
       "...         ...                 ...             ...        ...   \n",
       "31612     ZUFSP         NM_145062.1     NM_145062.1        NaN   \n",
       "252        ZW10         NM_004724.2     NM_004724.2        NaN   \n",
       "15920      ZXDB         NM_007157.3     NM_007157.3        NaN   \n",
       "24628    ZYG11B         NM_024646.1     NM_024646.1        NaN   \n",
       "28926     ZZEF1         NM_015113.3     NM_015113.3        NaN   \n",
       "\n",
       "       Entrez_Gene_ID  ...  Inf_149 Inf_152 Inf_157 Inf_159a Inf_162a  \\\n",
       "25103        346389.0  ...   7.0267  6.9027  6.9288   6.9657   7.0408   \n",
       "29049        339010.0  ...   7.0348  6.9117  7.0070   7.0041   7.1023   \n",
       "42813        728378.0  ...   7.0592  7.0157  7.0641   6.9264   7.0590   \n",
       "47175             2.0  ...   6.9565  7.0456  7.0070   7.0510   6.9608   \n",
       "7232         144568.0  ...   7.0741  7.1184  7.0750   7.0510   6.9889   \n",
       "...               ...  ...      ...     ...     ...      ...      ...   \n",
       "31612        221302.0  ...   8.1171  7.7784  7.7894   8.1040   8.1978   \n",
       "252            9183.0  ...   7.9831  8.1483  7.9581   7.2240   7.5738   \n",
       "15920        158586.0  ...   7.1934  7.1242  7.2487   7.0580   7.0641   \n",
       "24628         79699.0  ...   8.0346  8.6077  8.3911   8.4578   8.6710   \n",
       "28926         23140.0  ...   7.8563  8.4258  8.2854   7.7498   8.3944   \n",
       "\n",
       "       Inf_164 Inf_191  Inf_198 Inf_203 Fold change  \n",
       "25103   6.9972  6.9992   6.9750  7.0849         NaN  \n",
       "29049   7.0822  7.0845   7.0491  7.0978         NaN  \n",
       "42813   7.1122  7.0724   7.1493  7.0153         NaN  \n",
       "47175   7.0524  6.9637   6.9195  7.1239         NaN  \n",
       "7232    6.9835  7.0724   7.0491  7.0849         NaN  \n",
       "...        ...     ...      ...     ...         ...  \n",
       "31612   7.8296  7.7331   7.8357  7.8272         NaN  \n",
       "252     7.8844  7.7331   7.6729  7.5341     -0.3903  \n",
       "15920   7.1840  7.1504   7.3298  7.1376         NaN  \n",
       "24628   8.2491  8.1199   8.1083  8.6916         NaN  \n",
       "28926   8.3913  8.4547   8.1570  8.3454         NaN  \n",
       "\n",
       "[30804 rows x 93 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# identifying duplicate rows, and elimnating and replacing with max expression levels, across examples.\n",
    "\n",
    "# drop duplicate \n",
    "test_column = 'ILMN_Gene'\n",
    "full_data = raw_data.drop_duplicates(subset=['Probe_Id'], keep='first')\n",
    "\n",
    "# get boolean series of duplicates based on test_column name\n",
    "dd_data = full_data.duplicated(subset=[test_column], keep=False)\n",
    "\n",
    "# get index of the duplicates - i.e. the ones where True\n",
    "dup_index = dd_data[dd_data].index\n",
    "\n",
    "# create new dataframe, dropping all rows in the duplicate index\n",
    "clean_df = full_data.drop(index=dup_index)\n",
    "\n",
    "clean_df.sort_values(by=['ILMN_Gene'])\n",
    "# test to see if there are any further duplicate rows based on that test column\n",
    "#clean_dups = clean_df.duplicated(subset=['ILMN_Gene'], keep='first')\n",
    "#print(clean_dups)\n",
    "#clean_dups[clean_dups].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "relative-calendar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NuID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>Entrez_Gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Inf_149</th>\n",
       "      <th>Inf_152</th>\n",
       "      <th>Inf_157</th>\n",
       "      <th>Inf_159a</th>\n",
       "      <th>Inf_162a</th>\n",
       "      <th>Inf_164</th>\n",
       "      <th>Inf_191</th>\n",
       "      <th>Inf_198</th>\n",
       "      <th>Inf_203</th>\n",
       "      <th>Fold change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0_9dKJSRdZ6nRKsR5o</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_42795</td>\n",
       "      <td>ILMN_309061</td>\n",
       "      <td>DCDC2B</td>\n",
       "      <td>NM_001099434.1</td>\n",
       "      <td>NM_001099434.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>149069.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0826</td>\n",
       "      <td>7.0923</td>\n",
       "      <td>7.1350</td>\n",
       "      <td>6.9426</td>\n",
       "      <td>7.0748</td>\n",
       "      <td>6.9416</td>\n",
       "      <td>7.0845</td>\n",
       "      <td>7.0644</td>\n",
       "      <td>7.1239</td>\n",
       "      <td>0.0169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0_9Zd7UXqV4FRJ5XdE</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_35279</td>\n",
       "      <td>ILMN_35279</td>\n",
       "      <td>LOC649978</td>\n",
       "      <td>XM_939064.1</td>\n",
       "      <td>XM_939064.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>649978.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0010</td>\n",
       "      <td>7.1420</td>\n",
       "      <td>7.1856</td>\n",
       "      <td>7.1013</td>\n",
       "      <td>7.1069</td>\n",
       "      <td>7.0127</td>\n",
       "      <td>7.0489</td>\n",
       "      <td>7.1493</td>\n",
       "      <td>7.1517</td>\n",
       "      <td>0.0317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0_CoxIt3fv09UVwOo0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_85446</td>\n",
       "      <td>ILMN_85446</td>\n",
       "      <td>HS.302418</td>\n",
       "      <td>Hs.302418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.302418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0177</td>\n",
       "      <td>7.0861</td>\n",
       "      <td>7.0184</td>\n",
       "      <td>7.0380</td>\n",
       "      <td>6.9952</td>\n",
       "      <td>6.9835</td>\n",
       "      <td>7.0963</td>\n",
       "      <td>7.0961</td>\n",
       "      <td>7.0978</td>\n",
       "      <td>-0.0050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0_DHz7_krTURH3lVHk</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_26906</td>\n",
       "      <td>ILMN_26906</td>\n",
       "      <td>GIMAP4</td>\n",
       "      <td>NM_018326.2</td>\n",
       "      <td>NM_018326.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55303.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.1522</td>\n",
       "      <td>11.0532</td>\n",
       "      <td>12.8357</td>\n",
       "      <td>12.8616</td>\n",
       "      <td>11.9632</td>\n",
       "      <td>12.6279</td>\n",
       "      <td>10.6621</td>\n",
       "      <td>12.0104</td>\n",
       "      <td>12.0981</td>\n",
       "      <td>-0.6525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0_dJiKAH3o6kidUsOo</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_135414</td>\n",
       "      <td>ILMN_135414</td>\n",
       "      <td>HS.583233</td>\n",
       "      <td>Hs.583233</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.583233</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0665</td>\n",
       "      <td>6.9750</td>\n",
       "      <td>6.9514</td>\n",
       "      <td>7.0041</td>\n",
       "      <td>6.9754</td>\n",
       "      <td>7.1466</td>\n",
       "      <td>6.9992</td>\n",
       "      <td>6.9608</td>\n",
       "      <td>6.9240</td>\n",
       "      <td>-0.0128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48791</th>\n",
       "      <td>ZzoXZ08Fvp.Kjq6aYk</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_127410</td>\n",
       "      <td>ILMN_127410</td>\n",
       "      <td>HS.575229</td>\n",
       "      <td>Hs.575229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.575229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9830</td>\n",
       "      <td>7.1989</td>\n",
       "      <td>7.1112</td>\n",
       "      <td>7.6210</td>\n",
       "      <td>7.5831</td>\n",
       "      <td>7.1840</td>\n",
       "      <td>7.2971</td>\n",
       "      <td>7.1131</td>\n",
       "      <td>7.0041</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48792</th>\n",
       "      <td>ZzQeInuQpqo7l66ISI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_106946</td>\n",
       "      <td>ILMN_106946</td>\n",
       "      <td>HS.543340</td>\n",
       "      <td>Hs.543340</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.543340</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0348</td>\n",
       "      <td>7.1355</td>\n",
       "      <td>7.0298</td>\n",
       "      <td>7.0171</td>\n",
       "      <td>7.1800</td>\n",
       "      <td>7.0672</td>\n",
       "      <td>7.1234</td>\n",
       "      <td>6.9608</td>\n",
       "      <td>7.0733</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48794</th>\n",
       "      <td>ZZQvowA1A0rdICPUkk</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_24212</td>\n",
       "      <td>ILMN_24212</td>\n",
       "      <td>CASP8AP2</td>\n",
       "      <td>NM_012115.2</td>\n",
       "      <td>NM_012115.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9994.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0826</td>\n",
       "      <td>7.1113</td>\n",
       "      <td>7.0518</td>\n",
       "      <td>7.0041</td>\n",
       "      <td>7.0160</td>\n",
       "      <td>6.9972</td>\n",
       "      <td>7.1369</td>\n",
       "      <td>7.0338</td>\n",
       "      <td>7.0494</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48798</th>\n",
       "      <td>ZztzUERXUExGBETFFU</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_14674</td>\n",
       "      <td>ILMN_14674</td>\n",
       "      <td>NPY6R</td>\n",
       "      <td>NR_002713.1</td>\n",
       "      <td>NR_002713.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0010</td>\n",
       "      <td>7.0157</td>\n",
       "      <td>6.9841</td>\n",
       "      <td>6.9733</td>\n",
       "      <td>7.0813</td>\n",
       "      <td>7.0127</td>\n",
       "      <td>6.9264</td>\n",
       "      <td>7.0644</td>\n",
       "      <td>6.9361</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48801</th>\n",
       "      <td>ZzXXie6icAfD693B6w</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_104503</td>\n",
       "      <td>ILMN_104503</td>\n",
       "      <td>HS.539415</td>\n",
       "      <td>Hs.539415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.539415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0423</td>\n",
       "      <td>7.0072</td>\n",
       "      <td>7.0298</td>\n",
       "      <td>7.0441</td>\n",
       "      <td>7.0160</td>\n",
       "      <td>7.0254</td>\n",
       "      <td>7.0245</td>\n",
       "      <td>7.1131</td>\n",
       "      <td>7.0494</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30736 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     NuID       Species   Source   Search_Key   Transcript  \\\n",
       "1      0_9dKJSRdZ6nRKsR5o  Homo sapiens   RefSeq   ILMN_42795  ILMN_309061   \n",
       "2      0_9Zd7UXqV4FRJ5XdE  Homo sapiens   RefSeq   ILMN_35279   ILMN_35279   \n",
       "3      0_CoxIt3fv09UVwOo0  Homo sapiens  Unigene   ILMN_85446   ILMN_85446   \n",
       "4      0_DHz7_krTURH3lVHk  Homo sapiens   RefSeq   ILMN_26906   ILMN_26906   \n",
       "5      0_dJiKAH3o6kidUsOo  Homo sapiens  Unigene  ILMN_135414  ILMN_135414   \n",
       "...                   ...           ...      ...          ...          ...   \n",
       "48791  ZzoXZ08Fvp.Kjq6aYk  Homo sapiens  Unigene  ILMN_127410  ILMN_127410   \n",
       "48792  ZzQeInuQpqo7l66ISI  Homo sapiens  Unigene  ILMN_106946  ILMN_106946   \n",
       "48794  ZZQvowA1A0rdICPUkk  Homo sapiens   RefSeq   ILMN_24212   ILMN_24212   \n",
       "48798  ZztzUERXUExGBETFFU  Homo sapiens   RefSeq   ILMN_14674   ILMN_14674   \n",
       "48801  ZzXXie6icAfD693B6w  Homo sapiens  Unigene  ILMN_104503  ILMN_104503   \n",
       "\n",
       "       ILMN_Gene Source_Reference_ID       RefSeq_ID Unigene_ID  \\\n",
       "1         DCDC2B      NM_001099434.1  NM_001099434.1        NaN   \n",
       "2      LOC649978         XM_939064.1     XM_939064.1        NaN   \n",
       "3      HS.302418           Hs.302418             NaN  Hs.302418   \n",
       "4         GIMAP4         NM_018326.2     NM_018326.2        NaN   \n",
       "5      HS.583233           Hs.583233             NaN  Hs.583233   \n",
       "...          ...                 ...             ...        ...   \n",
       "48791  HS.575229           Hs.575229             NaN  Hs.575229   \n",
       "48792  HS.543340           Hs.543340             NaN  Hs.543340   \n",
       "48794   CASP8AP2         NM_012115.2     NM_012115.2        NaN   \n",
       "48798      NPY6R         NR_002713.1     NR_002713.1        NaN   \n",
       "48801  HS.539415           Hs.539415             NaN  Hs.539415   \n",
       "\n",
       "       Entrez_Gene_ID  ...  Inf_149  Inf_152  Inf_157 Inf_159a Inf_162a  \\\n",
       "1            149069.0  ...   7.0826   7.0923   7.1350   6.9426   7.0748   \n",
       "2            649978.0  ...   7.0010   7.1420   7.1856   7.1013   7.1069   \n",
       "3                 NaN  ...   7.0177   7.0861   7.0184   7.0380   6.9952   \n",
       "4             55303.0  ...  12.1522  11.0532  12.8357  12.8616  11.9632   \n",
       "5                 NaN  ...   7.0665   6.9750   6.9514   7.0041   6.9754   \n",
       "...               ...  ...      ...      ...      ...      ...      ...   \n",
       "48791             NaN  ...   6.9830   7.1989   7.1112   7.6210   7.5831   \n",
       "48792             NaN  ...   7.0348   7.1355   7.0298   7.0171   7.1800   \n",
       "48794          9994.0  ...   7.0826   7.1113   7.0518   7.0041   7.0160   \n",
       "48798             NaN  ...   7.0010   7.0157   6.9841   6.9733   7.0813   \n",
       "48801             NaN  ...   7.0423   7.0072   7.0298   7.0441   7.0160   \n",
       "\n",
       "       Inf_164  Inf_191  Inf_198  Inf_203 Fold change  \n",
       "1       6.9416   7.0845   7.0644   7.1239      0.0169  \n",
       "2       7.0127   7.0489   7.1493   7.1517      0.0317  \n",
       "3       6.9835   7.0963   7.0961   7.0978     -0.0050  \n",
       "4      12.6279  10.6621  12.0104  12.0981     -0.6525  \n",
       "5       7.1466   6.9992   6.9608   6.9240     -0.0128  \n",
       "...        ...      ...      ...      ...         ...  \n",
       "48791   7.1840   7.2971   7.1131   7.0041         NaN  \n",
       "48792   7.0672   7.1234   6.9608   7.0733         NaN  \n",
       "48794   6.9972   7.1369   7.0338   7.0494         NaN  \n",
       "48798   7.0127   6.9264   7.0644   6.9361         NaN  \n",
       "48801   7.0254   7.0245   7.1131   7.0494         NaN  \n",
       "\n",
       "[30736 rows x 93 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_dups = clean_df.dropna(subset=['Transcript']).duplicated(subset=['Transcript'], keep=False)\n",
    "dup_index_2 = clean_dups[clean_dups].index\n",
    "clean_df_2 = clean_df.drop(index=dup_index_2)\n",
    "clean_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fourth-month",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "crazy-correspondence",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
