---
title: "deseq2, unpaired all norm analysis"
author: "Robert Andrews"
date: "23/04/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

Attached is a an Excel workbook. Please look at the worksheet that is called ‘FOR ROB’. It contains the sequencing IDs you already have and a parallel column with all the treatments. (This is translated into targets.csv, which has the patient ID and their group). For your analyses and subsequent DEG comparisons (see below for breakdown) we want to make it clear what are the cases (C-LOS and P-LOS) and what are the controls (N-LOS, INDEX and contaminant). Peter has tried to make that clear in the treatment column.

Peter is requesting that the pre-filtering parameter should be done with >=10 reads?

When you carry out the comparisons Peter would like to see the PCA plots for vsd counts (not FPKM); Dendrograms or heat maps of similarity and MA plots (that have removed noise -- Peter thinks the command is plotMA(resLFC, ylim=c(-2,2))? Peter would also like to enquire whether we could request that an independent filtering also be applied for CV>=10%?

Groups are:
C-LOS - confirmed bacterial sepsis
P-lOS - possible sepsis / clinical sepsis
N-LOS - controls
INDEX - control, some paried to patients who subsequently became septic
Contaminant - infection deemed to be due to contamination on the line => consider as controls

The DEG contrasts we would like for direction case vs control are:

 C-LOS vs N-LOS
  P-LOS vs N-LOS
 INDEX vs N-LOS
 C-LOS vs (N-LOS + INDEX)
 (P-LOS + C-LOS) vs (N-LOS + INDEX)
 C-LOS vs (N-LOS + INDEX + contaminant)
 (C-LOS +P-LOS) vs (N-LOS + INDEX + contaminant)

Set working directory and define the required comparisons

```{r}

setwd("/Users/Ed/Documents/GitHub/sepsis_ml_omics_msc/dataset_pearth/an0304/bin")

#countInputFiles <- c("all.markdup.genecount", "all.rmdup.genecount")
countInputFiles <- c("all.markdup.genecount")

#Coefficient of Variation = (Standard Deviation / Mean) * 100.
coefficientOfVariationCutoff <- 10

fpkmCountfilters <- c("nofilter", 0, 0.5, 5, 10)
#fpkmCountfilters <- c(5)

myComparisons <- c("C-LOS_vs_Contaminant", "C-LOS_vs_N-LOS", "C-LOS_vs_P-LOS", "N-LOS_vs_Contaminant", "N-LOS_vs_C-LOS", "N-LOS_vs_P-LOS", "P-LOS_vs_Contaminant", "P-LOS_vs_N-LOS", "P-LOS_vs_C-LOS", "Contaminant_vs_C-LOS", "Contaminant_vs_N-LOS", "Contaminant_vs_P-LOS")

apadjCutoffs <- c("allgenes", "top50genes", "sig0-01", "sig0-05")

#outliers <- NULL

## reasons for removing = low yield
## note low run are retained, because taking them out gives no significant genes ("11289_ICG200462A", "81051_SOG1000041A")


outliers <- NULL

## first-round
#outliers <- c("10481_sepsis", "10108_sepsis", "10501_Index", "10127_sepsis", "10281_Index", "10509_Index")

pcaWindows <- c("sampleGroup") 


```

Load relevant libraries

```{r}

library(DESeq2)     # differential analysis

library(ape)        # clustering QC
library(ggplot2)    # PCA plotting
library(ggrepel)

library(dplyr)      # data manipulations
library(docstring)  # docstring information

```

Create the uber input counts files

Read in all data

```{r}

## get top, middle and bottom

rawDataTop <- read.table("../input/top/all.markdup.genecount.txt", sep="\t", header=T, check.names=F)
rawDataMiddle <- read.table("../input/middle/all.markdup.genecount.txt", sep="\t", header=T, check.names=F)
rawDataBottom <- read.table("../input/bottom/all.markdup.genecount.txt", sep="\t", header=T, check.names=F)

## get the topups

rawDataTopups <- read.table("../input/topups/all.markdup.genecount.txt", sep="\t", header=T, check.names=F)

```

merge top, middle and bottom

```{r}

rawData <- merge(rawDataTop, rawDataMiddle, by="ensemblGeneID")

rawData <-  dplyr::select(rawData, -geneLength.y, -geneName.y, -geneBiotype.y)
colnames(rawData)[2:4] <- c("geneLength", "geneName", "geneBiotype")

rawData <- merge(rawData, rawDataBottom, by="ensemblGeneID")

rawData <-  dplyr::select(rawData, -geneLength.y, -geneName.y, -geneBiotype.y)
colnames(rawData)[2:4] <- c("geneLength", "geneName", "geneBiotype")

```

merge top-ups into raw data

```{r}

## get a list of merged samples - empty vector

replacedups <- c()

## paranoia check: are the order of genes the same in the tops and the top-ups??

if (all(rawData$ensemblGeneID != rawDataTopups$ensemblGeneID)) { stop() }

mergedRawData <- rawData

for (coln in colnames(dplyr::select(rawDataTopups, -ensemblGeneID, -geneLength, -geneName, -geneBiotype))) {
  
  # identify the matching column names in the merged raw data by string replacement as topup columns named differently
  oldColn <- gsub(".b", "", coln)
  oldColn <- gsub("NO_LOS", "NOLOS", oldColn)
  
  if (oldColn %in% colnames(mergedRawData)) {
    
    mergedRawData[oldColn] <- mergedRawData[oldColn] + rawDataTopups[coln]
    replacedups <- c(replacedups, oldColn)

  }
  
}

```

Check that all top-ups have been merged

```{r}

topupColnames <- gsub("NO_LOS", "NOLOS", gsub(".b", "", colnames(dplyr::select(rawDataTopups, -ensemblGeneID, -geneLength, -geneName, -geneBiotype))))

if (all(topupColnames != replacedups)) { stop() }

```

Write the raw counts to a master file in the uber input folder

```{r}

write.table(mergedRawData, file="../input/uber/all.markdup.genecount.txt", sep="\t", row.names = F, col.names = T, quote = F)

geneLengths <- dplyr::select(rawData, ensemblGeneID, geneLength)

```

======== end of merging and deduplicating of datasets ============

Define function used to run the differential expression analysis

```{r}


RunDeseq2 <- function(rc, gl) {
    
    #' Generate analysis object
    #'
    #' Creates analysis object from table of raw counts and table of gene lengths
    #' @param rc The table of raw counts with ensembl_gene_ID as row index and patient sampel ID as column name
    #' @param gl The table of gene lengths
    
    ## create the experimental design object for DeSeq2.
    ## creates a dataframe with mapping of the sample id to the sample group
    
    exptDesign <- data.frame(
            row.names = colnames(rc),
            condition = sampleGroups)
    
    ## construct experiment object -object used to store input values, intermediate calculations and result of differential expression analysis
    ## takes as input the raw counts (pre-processed data), mapping of samples to groups, colData - rows correspond to the columns of the raw counts
    ## design specifies how the counts in countData depend on the variables in colData
    ## effectively does an ANOVA and looks at the impact of group membership on differential expression
    
    exptObject <- DESeqDataSetFromMatrix(
          countData = rc,
          colData = exptDesign,
          design = ~ condition)

    ## run the analysis
    ## returns matrix of ensembl gene ids with baseMean, log2foldchange, lfcSE, stat, p-value, padj
    ## this DESeq does a normalisation, generates scaling factors, scales by the total number of reads in each of the columns - READ DOCUMENTATION
    ## aim of the normalisation, puts each column (sample) on the same scale - scaling factor for the column
    ## different to microarrays that do a quantile normalisation
    ## if you run the results function without any other arguments, it runs an anova, and you get a list of the most significant genes
    ## if you run the results function with an argument (coefficient), e.g. P-Los, c-los, then you get a pairwise comparison
    ## to compare healthy vs. control, take the targets.csv and add another column and refernece in the exptDesign object
    analysisObject = DESeq(exptObject)

    ## add gene lengths to the analysis object for the FPKM calculation
    
    ## ensure that the ensembl gene ids match first
    geneLengthsTemp <- gl[gl$ensemblGeneID %in% rownames(analysisObject),]
    
    ## add in a $basepairs column to the analysis Object output - required for FPKM normalisation based on gene length
    mcols(analysisObject)$basepairs <- geneLengthsTemp[match(rownames(analysisObject), geneLengthsTemp$ensemblGeneID),]$geneLength
    
    return(analysisObject)    
}

```


Perform the differential analysis over the count filters and input data files

```{r}

## loop over the 2 files under input/

for (countfilter in fpkmCountfilters) {
  
  for (file in countInputFiles) {
    
    ## READ INPUT FILES 
    
    ## read targets file
    targets <- read.table("../resources/targets.csv", sep=",", header=T)
   
    ## read input data file
    
    rawCounts <- read.table(paste("../input/uber/", file, ".txt", sep=""), sep="\t", header=TRUE, check.names=F)
    
    ## INPUT DATA PRE-PROCESSING
    
    ## remove outliers from input files
    
    if (length(outliers) > 0) {
      
      # remove any rows from the targets table, where the sample id is in the outliers vector 
      targets <- targets[!targets$analysisID %in% outliers, ]
      
      # remove any columns from the rawCounts table (i.e. the merged raw data) where sample id is in the outliers vector
      rawCounts <- rawCounts[, !(names(rawCounts) %in% outliers)]
    }
    
    ## remove any minus sign from the column names (not the case in current dataset)
    
    colnames(rawCounts) <- sub("-", ".", colnames(rawCounts))
    
    # parse out the 3 first columns of annotation
    
    annotationNames <- rawCounts[,c(1:4)]
    
    # get the sample groups from targets
    
    sampleGroups <- targets$sampleGroup

    # add rownames to the raw data using ensembleGeneID as the index, and remove all four annotation columns
    
    rownames(rawCounts) = rawCounts[,1]
    rawCounts[,c(1:4)] <- NULL
    
    ## check the order of the targets files and input counts file are identical
    
    if (!identical(as.character(targets$analysisID), colnames(rawCounts))) {
      rawCounts <- dplyr::select(rawCounts, as.character(targets$analysisID))
    }
    
        
    ## RUN DEseq ANALYSIS WITHOUT FILTERING GENES
    
    analysisObject <- RunDeseq2(rawCounts, geneLengths)
      
    ## extract raw, scaling factor normalised and fpkm normalised counts from the output, and check rows match
    
    rawCounts <- counts(analysisObject, normalized = FALSE)
    normalisedCounts <- counts(analysisObject, normalized = TRUE)
    fpkmNormalisedCounts <- fpkm(analysisObject, robust = TRUE)
    
    if (!identical(rownames(rawCounts), rownames(normalisedCounts))) { stop() }
    if (!identical(rownames(fpkmNormalisedCounts), rownames(rawCounts))) { stop() }
    
    ## Filter genes based on minimum coefficient of variation over entire gene - exclude any gene where cov is less than threshold
    
    ## Coefficient of Variation = (Standard Deviation / Mean) * 100.
    ## https://www.statology.org/coefficient-of-variation-in-r/
    
    cov <- apply(fpkmNormalisedCounts, 1, function(x) sd(x) / mean(x) * 100)
    
    rawCountsToKeep <- rawCounts[cov >= coefficientOfVariationCutoff, ]
    
    ## filter the object by minimum fpkmNormalisedCounts then re run the DEseq on filtered data
    
    if (countfilter != "nofilter") {
      
      ## exclude any gene where the fpkm normalised counts don't meet a minimum threshold for at least one gene. (note <0.5 indicates unclear if gene present)
    
      rawCountsToKeep <- rawCounts[rowSums(fpkmNormalisedCounts > countfilter) >= 1, ]
    
      ## run analysis with filters
    
      analysisObject <- RunDeseq2(rawCountsToKeep, geneLengths)
    }
    
    ## extract raw, scaling factor normalised and fpkm normalised counts from the filtered output, and check rows match
    
    rawCounts <- counts(analysisObject, normalized = FALSE)
    normalisedCounts <- counts(analysisObject, normalized = TRUE)
    fpkmNormalisedCounts <- fpkm(analysisObject, robust = TRUE)
    
    if (!identical(rownames(rawCounts), rownames(normalisedCounts))) { stop() }
    if (!identical(rownames(fpkmNormalisedCounts), rownames(rawCounts))) { stop() }
    
    ## merge raw and normalised data
    ## add the words "raw.counts" to each column names (ref for R regex: https://colauttilab.github.io/RCrashCourse/4_regex.html)
    
    colnames(rawCounts) = gsub("^", "raw.counts.", colnames(rawCounts))
    colnames(normalisedCounts) = gsub("^", "norm.counts.", colnames(normalisedCounts))
    colnames(fpkmNormalisedCounts) = gsub("^", "fpkm.norm.counts.", colnames(fpkmNormalisedCounts))
    
    tempData = merge(rawCounts, normalisedCounts, by="row.names", all=T)
    rownames(tempData) <- tempData[,1]
    tempData[,1] <- NULL
  
    ## putting everything into one file
    
    finalData = merge(tempData, fpkmNormalisedCounts, by="row.names", all=T)
    rownames(finalData) <- finalData[,1]
    finalData[,1] <- NULL
  
    printNormCounts <- data.frame(ensembleID=rownames(fpkmNormalisedCounts), fpkmNormalisedCounts)
    
    ## circle over the comparisons of everything versus everything
     
    for (myComparison in myComparisons) {
      
        numerator <- unlist(strsplit(myComparison, "_vs_"))[1]
        denominator <- unlist(strsplit(myComparison, "_vs_"))[2]
        
        print(myComparison)
                            
        result <- as.data.frame(results(analysisObject, contrast=c("condition", numerator, denominator), independentFiltering=TRUE, pAdjustMethod="BH"))
        
        slimData <- as.data.frame(cbind(rownames(finalData), finalData))
                       
        colnames(slimData)[1] <-  "tracking_id"
             
         ## get only the genes of interest
        
        slimData <- dplyr::select(slimData, tracking_id, colnames(dplyr::select(slimData, paste("raw.counts", unlist(subset(targets, sampleGroup == numerator, analysisID)), sep="."), paste("raw.counts", unlist(subset(targets, sampleGroup == denominator, analysisID)), sep="."), paste("norm.counts", unlist(subset(targets, sampleGroup == numerator, analysisID)), sep="."), paste("norm.counts", unlist(subset(targets, sampleGroup == denominator, analysisID)), sep="."), paste("fpkm.norm.counts", unlist(subset(targets, sampleGroup == numerator, analysisID)), sep="."), paste("fpkm.norm.counts", unlist(subset(targets, sampleGroup == denominator, analysisID)), sep="."))))
                          
    
        printData <- merge(slimData, result, by.x = "tracking_id", by.y = "row.names", all = T)

        ## add annotation
            
        printData <- merge(annotationNames, printData, by.x="ensemblGeneID", by.y="tracking_id") 
                
        printData <- printData[order(printData$pvalue), ]    
        
        ## add second header
        
        printData %>% mutate_if(is.factor, as.character) -> printData2
        printData2 <- rbind(colnames(printData), printData2)
        
        ## get group names for final output file
    
        temp <- colnames(slimData[-1])
        temp <- gsub("^raw.counts.", "", gsub("^norm.counts.", "", gsub("^fpkm.norm.counts.", "", temp)))
        
        outputColHeadings <- c(colnames(annotationNames), as.vector(targets$sampleGroup[match(temp, targets$analysisID)]), colnames(result))
  
        printData2 <- rbind(outputColHeadings, printData2)
        
        ## comment on outliers removed
        commentLine <- rep("", length(outputColHeadings))
        commentLine[1] <- paste("## Outliers removed: ", paste(outliers, collapse=" "), sep="")
        printData2 <- rbind(commentLine, printData2)
        
        write.table(printData2, file=paste("../output/", numerator, "_vs_", denominator, ".", file, ".FPKM", countfilter, ".allnorm.txt", sep=""), col.names=F, row.names=F, sep="\t", quote=F)
                    
     }
  }
}

```

================ ANOVA and PCA ======================

Define functions used in the code below

```{r}

GetUniqueGenes <- function(anr, ty) {
  
    #' Generate a list of significant genes based on a adjusted p-value filter
    #'
    #' xx
    #' 
    #' @param anr data matrix that you get at the end - one line per - the final data file
    #' @param ty Adjusted p cut offs
  
  sgenes <- NULL
  anr <- anr[order(anr$pvalue), ]  
  
  if (ty == "allgenes") {
     sgenes <- rownames(anr)
  } else if (grepl("^sig", ty)) {
    ty <- gsub("-", ".", gsub("sig", "", ty))
    
    ## need to get genes that are also NA but significant
    
    for (i in 1:length(anr[,1])) {
    
      if (is.na(anr[i,]$padj)) {
        sgenes <- c(sgenes, rownames((anr[i,]))) 
      } else if ((as.numeric(anr[i,]$padj) < as.numeric(ty))) {
        sgenes <- c(sgenes, rownames((anr[i,]))) 
      } else {
        break
      }
    }
  } else if (grepl("^top", ty)) {
    ty <- gsub("top", "", gsub("genes", "", ty))
    sgenes <- rownames(anr[1:as.numeric(ty),])
  } else {
    stop("cuttoff not recognised")
  }
  
  return (sgenes)
}
```

Perform ANOVA and PCA analysis using the differentially expressed values over all sample groups

```{r}
## perform ANOVA for the PCA: looking for changing genes across the entire experiment
## NOTE: this is currently written for a single analysisObject - need to reinstate loops over multiple analysis objects    

    anovaResult <- as.data.frame(results(analysisObject, pAdjustMethod="BH"))
    anovaResult <- anovaResult[order(anovaResult$pvalue), ]  
    
    ## plot all genes PCA
    
    for (dirName in apadjCutoffs) {
      
      print(dirName )
      
      ## get unique genes
      uniqueSignificantGenes <- GetUniqueGenes(anovaResult, dirName);
      
      if (length(uniqueSignificantGenes) > 5) {
        
        ## create a directory under output if not there already
      
        if (!file.exists(paste("../output/", dirName, sep=""))) {
          dir.create(file.path("../output/", dirName))
        }
        
        
        ## create subset of significant genes
        
        pcaDD <- subset(fpkmNormalisedCounts, rownames(rawCounts) %in% uniqueSignificantGenes)
        
        
        ## create object for hierarchical clustering
        
        hierData <- pcaDD
    
        ## change label name to feature sampleGroup
        b <- NULL
    
        for (a in gsub("^fpkm.norm.counts.", "", colnames(hierData))) {
          b <- c(b, paste(as.vector(targets$sampleGroup[targets$analysisID == a]), a, sep=" "))
        }
    
        colnames(hierData) <- b
    
        d <- cor(hierData, method="pearson")
        hc <- hclust(dist(1-d))
    
        ## write to outputfile
    
        hierData <- data.frame(ensembleGeneID=rownames(hierData), hierData)
        
        aa <- subset(rawData, ensemblGeneID %in% rownames(hierData), select=c(ensemblGeneID, geneName))
        bb <- merge(aa, hierData,  by.y="row.names", by.x="ensemblGeneID")
        bb[,3] <- NULL
                   
        write.table(bb, file=paste("../output/", dirName, "/hcluster.", file, ".FPKM", countfilter, ".FPKM.", dirName, ".txt", sep=""), col.names=T, row.names=F, sep="\t", quote=F)
    
    
        ## create pca object
    
        pca <- prcomp(t(pcaDD), center=TRUE, scale=F)
        scores <- data.frame(targets$analysisID, pca$x[,1:2])
    
        # write PCA coordinates to file
    
        write.table(scores, paste("../output/", dirName, "/PCA.", file, ".FPKM", countfilter, ".coord.", dirName, ".txt", sep=""), sep="\t", row.names = F, quote = F)
       
        # write loadings to a file
       
        aload <- abs(pca$rotation)
        loadings <- sweep(aload, 2, colSums(aload), "/")
        loadings <- cbind(row.names(loadings), loadings)
        write.table(loadings, paste("../output/", dirName, "/PCA.", file, ".FPKM", countfilter, ".load.", dirName, ".txt", sep=""), sep="\t", row.names = F, quote = F)
              
        # write percentage contributions to file
       
        write.table(summary(pca)$importance[2,], paste("../output/", dirName, "/PCA.", file, ".FPKM", countfilter, ".perc.", dirName, ".txt", sep=""), sep="\t", row.names = T, quote = F)
       
        ## open the PDF
      
        pointSize <- 0.4
    
        pdf(paste("../output/", dirName, "/PCA.", file, ".FPKM", countfilter, ".", dirName, ".pdf", sep=""), onefile=T)
    
        ## print dendrogram
        
        tiplabs <- as.phylo(hc)$tip.label
        numbertiplabs<-length(tiplabs)
        colourtips<-rep("blue",numbertiplabs)
        colourtips[grep("CTR2",tiplabs)]<-"red"

        plot.phylo(as.phylo(hc), type="p", show.node.label=TRUE, main=paste(file, "plotting FPKMs using [", length(uniqueSignificantGenes), "] genes.  Outliers removed are: [" , paste(outliers, collapse = " "), "]", sep=" "), cex=pointSize, cex.main=pointSize, tip.color=colourtips)
        
        ## print PCA windows
        
        for (x in pcaWindows) {
          print(qplot(x=PC1, y=PC2, main = paste(file, "plotting FPKMs using [", length(uniqueSignificantGenes), "] genes.  Outliers removed are: [" , paste(outliers, collapse = " "), "]", sep=" "), data=scores, label=factor(targets$analysisID), colour=factor(targets[[x]])) + geom_text_repel(size=2.5) +  scale_colour_discrete(name=x) + theme(plot.title = element_text(size = 5))  )
        }
    
        dev.off()
      }
    }
#  }
#}

```

