{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "anonymous-tourism",
   "metadata": {},
   "source": [
    "# leo adults dataset: pre-processing for ml input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cheap-garbage",
   "metadata": {},
   "source": [
    "### Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "hydraulic-washer",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import stats\n",
    "from collections import OrderedDict\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import warnings\n",
    "from joblib import dump, load\n",
    "from pickle import dump, load\n",
    "warnings.filterwarnings(action=\"ignore\", message=\"^internal gelsd\")\n",
    "#np.random.seed(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "behavioral-civilization",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/Ed/Documents/GitHub/sepsis_ml_omics_msc'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lesser-connectivity",
   "metadata": {},
   "source": [
    "### Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "handed-ebony",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Normalised counts data shape:  (22011, 106)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GSM1914807</th>\n",
       "      <th>GSM1914808</th>\n",
       "      <th>GSM1914809</th>\n",
       "      <th>GSM1914810</th>\n",
       "      <th>GSM1914811</th>\n",
       "      <th>GSM1914812</th>\n",
       "      <th>GSM1914813</th>\n",
       "      <th>GSM1914814</th>\n",
       "      <th>GSM1914815</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1914902</th>\n",
       "      <th>GSM1914903</th>\n",
       "      <th>GSM1914904</th>\n",
       "      <th>GSM1914905</th>\n",
       "      <th>GSM1914906</th>\n",
       "      <th>GSM1914907</th>\n",
       "      <th>GSM1914908</th>\n",
       "      <th>GSM1914909</th>\n",
       "      <th>GSM1914910</th>\n",
       "      <th>GSM1914911</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2315554</td>\n",
       "      <td>7.988658</td>\n",
       "      <td>8.437184</td>\n",
       "      <td>8.277160</td>\n",
       "      <td>8.279856</td>\n",
       "      <td>8.124447</td>\n",
       "      <td>7.888959</td>\n",
       "      <td>7.743694</td>\n",
       "      <td>7.949590</td>\n",
       "      <td>8.246015</td>\n",
       "      <td>...</td>\n",
       "      <td>8.175743</td>\n",
       "      <td>8.478647</td>\n",
       "      <td>8.522634</td>\n",
       "      <td>8.267206</td>\n",
       "      <td>8.245167</td>\n",
       "      <td>8.314857</td>\n",
       "      <td>8.206667</td>\n",
       "      <td>8.073364</td>\n",
       "      <td>8.160501</td>\n",
       "      <td>8.149364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2315633</td>\n",
       "      <td>7.604336</td>\n",
       "      <td>7.877958</td>\n",
       "      <td>7.898660</td>\n",
       "      <td>7.657952</td>\n",
       "      <td>7.865661</td>\n",
       "      <td>7.341377</td>\n",
       "      <td>7.465072</td>\n",
       "      <td>7.701582</td>\n",
       "      <td>7.645253</td>\n",
       "      <td>...</td>\n",
       "      <td>7.617291</td>\n",
       "      <td>7.846885</td>\n",
       "      <td>7.734627</td>\n",
       "      <td>7.723670</td>\n",
       "      <td>7.885184</td>\n",
       "      <td>7.709711</td>\n",
       "      <td>7.809246</td>\n",
       "      <td>7.773138</td>\n",
       "      <td>7.562412</td>\n",
       "      <td>7.789600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2315674</td>\n",
       "      <td>7.756030</td>\n",
       "      <td>8.334455</td>\n",
       "      <td>8.129139</td>\n",
       "      <td>7.949146</td>\n",
       "      <td>8.140260</td>\n",
       "      <td>7.552136</td>\n",
       "      <td>7.662208</td>\n",
       "      <td>7.943859</td>\n",
       "      <td>7.948027</td>\n",
       "      <td>...</td>\n",
       "      <td>7.941620</td>\n",
       "      <td>8.223983</td>\n",
       "      <td>8.157951</td>\n",
       "      <td>8.016826</td>\n",
       "      <td>7.979801</td>\n",
       "      <td>8.059808</td>\n",
       "      <td>8.092450</td>\n",
       "      <td>7.909820</td>\n",
       "      <td>8.057985</td>\n",
       "      <td>8.111045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2315739</td>\n",
       "      <td>7.742156</td>\n",
       "      <td>8.070674</td>\n",
       "      <td>7.978276</td>\n",
       "      <td>7.688884</td>\n",
       "      <td>8.048886</td>\n",
       "      <td>7.553671</td>\n",
       "      <td>7.618928</td>\n",
       "      <td>7.896778</td>\n",
       "      <td>7.711746</td>\n",
       "      <td>...</td>\n",
       "      <td>7.752970</td>\n",
       "      <td>7.872649</td>\n",
       "      <td>7.963431</td>\n",
       "      <td>7.957844</td>\n",
       "      <td>7.752404</td>\n",
       "      <td>7.978197</td>\n",
       "      <td>8.019349</td>\n",
       "      <td>7.808959</td>\n",
       "      <td>7.879470</td>\n",
       "      <td>7.684881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2315894</td>\n",
       "      <td>9.535384</td>\n",
       "      <td>10.219614</td>\n",
       "      <td>10.075132</td>\n",
       "      <td>9.614065</td>\n",
       "      <td>10.015771</td>\n",
       "      <td>9.765352</td>\n",
       "      <td>9.599523</td>\n",
       "      <td>10.154484</td>\n",
       "      <td>9.988529</td>\n",
       "      <td>...</td>\n",
       "      <td>10.007208</td>\n",
       "      <td>10.178184</td>\n",
       "      <td>10.283674</td>\n",
       "      <td>10.045307</td>\n",
       "      <td>9.725878</td>\n",
       "      <td>9.945120</td>\n",
       "      <td>10.107392</td>\n",
       "      <td>10.064496</td>\n",
       "      <td>9.860088</td>\n",
       "      <td>10.097237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 106 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  GSM1914807  GSM1914808  GSM1914809  GSM1914810  GSM1914811  \\\n",
       "0     2315554    7.988658    8.437184    8.277160    8.279856    8.124447   \n",
       "1     2315633    7.604336    7.877958    7.898660    7.657952    7.865661   \n",
       "2     2315674    7.756030    8.334455    8.129139    7.949146    8.140260   \n",
       "3     2315739    7.742156    8.070674    7.978276    7.688884    8.048886   \n",
       "4     2315894    9.535384   10.219614   10.075132    9.614065   10.015771   \n",
       "\n",
       "   GSM1914812  GSM1914813  GSM1914814  GSM1914815  ...  GSM1914902  \\\n",
       "0    7.888959    7.743694    7.949590    8.246015  ...    8.175743   \n",
       "1    7.341377    7.465072    7.701582    7.645253  ...    7.617291   \n",
       "2    7.552136    7.662208    7.943859    7.948027  ...    7.941620   \n",
       "3    7.553671    7.618928    7.896778    7.711746  ...    7.752970   \n",
       "4    9.765352    9.599523   10.154484    9.988529  ...   10.007208   \n",
       "\n",
       "   GSM1914903  GSM1914904  GSM1914905  GSM1914906  GSM1914907  GSM1914908  \\\n",
       "0    8.478647    8.522634    8.267206    8.245167    8.314857    8.206667   \n",
       "1    7.846885    7.734627    7.723670    7.885184    7.709711    7.809246   \n",
       "2    8.223983    8.157951    8.016826    7.979801    8.059808    8.092450   \n",
       "3    7.872649    7.963431    7.957844    7.752404    7.978197    8.019349   \n",
       "4   10.178184   10.283674   10.045307    9.725878    9.945120   10.107392   \n",
       "\n",
       "   GSM1914909  GSM1914910  GSM1914911  \n",
       "0    8.073364    8.160501    8.149364  \n",
       "1    7.773138    7.562412    7.789600  \n",
       "2    7.909820    8.057985    8.111045  \n",
       "3    7.808959    7.879470    7.684881  \n",
       "4   10.064496    9.860088   10.097237  \n",
       "\n",
       "[5 rows x 106 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import and check data\n",
    "\n",
    "# alternative X matrices\n",
    "normalisedCounts = pd.read_csv('dataset_leo/output/normalised_exp.csv') #, index_col=0)\n",
    "\n",
    "# annotation and class labels\n",
    "labels_data = pd.read_csv('dataset_leo/output/targets.csv', index_col=0)\n",
    "\n",
    "# check\n",
    "print('Normalised counts data shape:  '+str(normalisedCounts.shape))\n",
    "normalisedCounts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "raised-mustang",
   "metadata": {},
   "source": [
    "### Pre-process X matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "nutritional-racing",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pre-processing function for all the perth datasets types\n",
    "\n",
    "def pre_process_X(raw_data):\n",
    "    \n",
    "    ''' dataframe -> datafram\n",
    "    Input dataframe of normalised rnaseq data, transformed and z-scored\n",
    "    '''\n",
    "    # rename first column to ensembleGeneID and drop any duplicate rows based on this column\n",
    "    df = raw_data.rename({\"Unnamed: 0\": \"affy_probe_ID\"}, axis=1).drop_duplicates(subset=[\"affy_probe_ID\"], keep='first')\n",
    "    \n",
    "    # transpose the dataframe so genes are the columns\n",
    "    df = df.transpose()\n",
    "    \n",
    "    # set the column names to first row, i.e ensembleGeneID, in the transposed frame\n",
    "    df = df.rename(columns=df.iloc[0]).drop(df.index[0])\n",
    "    \n",
    "    # drop columns containing any NaN values\n",
    "    nan_columns = df.columns[df.isna().any()].tolist()\n",
    "    df = df.drop(nan_columns, axis=1)\n",
    "    \n",
    "    # z-score the data\n",
    "    scaler = StandardScaler()\n",
    "    df_scaled = scaler.fit_transform(df)\n",
    "    df = pd.DataFrame(df_scaled, columns=df.columns, index=df.index)\n",
    "    \n",
    "    return df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "seventh-appreciation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2315554.0</th>\n",
       "      <th>2315633.0</th>\n",
       "      <th>2315674.0</th>\n",
       "      <th>2315739.0</th>\n",
       "      <th>2315894.0</th>\n",
       "      <th>2315918.0</th>\n",
       "      <th>2315951.0</th>\n",
       "      <th>2316218.0</th>\n",
       "      <th>2316245.0</th>\n",
       "      <th>2316379.0</th>\n",
       "      <th>...</th>\n",
       "      <th>4133728.0</th>\n",
       "      <th>4134740.0</th>\n",
       "      <th>7385511.0</th>\n",
       "      <th>7385515.0</th>\n",
       "      <th>7385547.0</th>\n",
       "      <th>7385552.0</th>\n",
       "      <th>7385611.0</th>\n",
       "      <th>7385641.0</th>\n",
       "      <th>7385683.0</th>\n",
       "      <th>7385696.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1914807</th>\n",
       "      <td>-0.801101</td>\n",
       "      <td>-0.477595</td>\n",
       "      <td>-1.261832</td>\n",
       "      <td>-0.814735</td>\n",
       "      <td>-1.697615</td>\n",
       "      <td>0.766582</td>\n",
       "      <td>-1.375160</td>\n",
       "      <td>-0.517497</td>\n",
       "      <td>-1.036075</td>\n",
       "      <td>-0.917664</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.193302</td>\n",
       "      <td>0.321372</td>\n",
       "      <td>-0.090157</td>\n",
       "      <td>0.617512</td>\n",
       "      <td>-0.645097</td>\n",
       "      <td>-0.851523</td>\n",
       "      <td>-1.100261</td>\n",
       "      <td>-0.583777</td>\n",
       "      <td>0.334289</td>\n",
       "      <td>-1.151476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1914808</th>\n",
       "      <td>1.225424</td>\n",
       "      <td>1.175072</td>\n",
       "      <td>1.800243</td>\n",
       "      <td>1.299245</td>\n",
       "      <td>0.604285</td>\n",
       "      <td>1.402227</td>\n",
       "      <td>1.150008</td>\n",
       "      <td>1.162674</td>\n",
       "      <td>-1.181174</td>\n",
       "      <td>-0.849105</td>\n",
       "      <td>...</td>\n",
       "      <td>0.494821</td>\n",
       "      <td>-0.699605</td>\n",
       "      <td>1.423479</td>\n",
       "      <td>0.468561</td>\n",
       "      <td>0.082516</td>\n",
       "      <td>-0.001238</td>\n",
       "      <td>-1.229338</td>\n",
       "      <td>1.906317</td>\n",
       "      <td>-1.284613</td>\n",
       "      <td>-1.936364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1914809</th>\n",
       "      <td>0.502405</td>\n",
       "      <td>1.300111</td>\n",
       "      <td>0.713339</td>\n",
       "      <td>0.704670</td>\n",
       "      <td>0.118215</td>\n",
       "      <td>1.387755</td>\n",
       "      <td>0.409928</td>\n",
       "      <td>1.446316</td>\n",
       "      <td>-0.059921</td>\n",
       "      <td>-0.200042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.734641</td>\n",
       "      <td>0.214251</td>\n",
       "      <td>-0.027840</td>\n",
       "      <td>0.457613</td>\n",
       "      <td>-0.463707</td>\n",
       "      <td>-0.587461</td>\n",
       "      <td>-0.553096</td>\n",
       "      <td>1.469179</td>\n",
       "      <td>0.052834</td>\n",
       "      <td>-1.200240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1914810</th>\n",
       "      <td>0.514587</td>\n",
       "      <td>-0.153755</td>\n",
       "      <td>-0.239512</td>\n",
       "      <td>-1.157534</td>\n",
       "      <td>-1.432917</td>\n",
       "      <td>1.173428</td>\n",
       "      <td>-1.339684</td>\n",
       "      <td>0.622747</td>\n",
       "      <td>-1.637608</td>\n",
       "      <td>-1.982493</td>\n",
       "      <td>...</td>\n",
       "      <td>0.344792</td>\n",
       "      <td>-0.812398</td>\n",
       "      <td>0.184977</td>\n",
       "      <td>0.633738</td>\n",
       "      <td>1.122524</td>\n",
       "      <td>-0.478908</td>\n",
       "      <td>-1.820808</td>\n",
       "      <td>0.285067</td>\n",
       "      <td>3.199836</td>\n",
       "      <td>-0.330872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1914811</th>\n",
       "      <td>-0.187580</td>\n",
       "      <td>1.100799</td>\n",
       "      <td>0.772210</td>\n",
       "      <td>1.159036</td>\n",
       "      <td>-0.081487</td>\n",
       "      <td>-1.336768</td>\n",
       "      <td>1.360797</td>\n",
       "      <td>-0.846084</td>\n",
       "      <td>-0.193670</td>\n",
       "      <td>0.307112</td>\n",
       "      <td>...</td>\n",
       "      <td>1.175015</td>\n",
       "      <td>0.732782</td>\n",
       "      <td>0.310884</td>\n",
       "      <td>-1.773219</td>\n",
       "      <td>-0.204938</td>\n",
       "      <td>1.053665</td>\n",
       "      <td>1.811451</td>\n",
       "      <td>0.039464</td>\n",
       "      <td>-1.424448</td>\n",
       "      <td>0.255503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22011 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            2315554.0  2315633.0  2315674.0  2315739.0  2315894.0  2315918.0  \\\n",
       "GSM1914807  -0.801101  -0.477595  -1.261832  -0.814735  -1.697615   0.766582   \n",
       "GSM1914808   1.225424   1.175072   1.800243   1.299245   0.604285   1.402227   \n",
       "GSM1914809   0.502405   1.300111   0.713339   0.704670   0.118215   1.387755   \n",
       "GSM1914810   0.514587  -0.153755  -0.239512  -1.157534  -1.432917   1.173428   \n",
       "GSM1914811  -0.187580   1.100799   0.772210   1.159036  -0.081487  -1.336768   \n",
       "\n",
       "            2315951.0  2316218.0  2316245.0  2316379.0  ...  4133728.0  \\\n",
       "GSM1914807  -1.375160  -0.517497  -1.036075  -0.917664  ...  -1.193302   \n",
       "GSM1914808   1.150008   1.162674  -1.181174  -0.849105  ...   0.494821   \n",
       "GSM1914809   0.409928   1.446316  -0.059921  -0.200042  ...   0.734641   \n",
       "GSM1914810  -1.339684   0.622747  -1.637608  -1.982493  ...   0.344792   \n",
       "GSM1914811   1.360797  -0.846084  -0.193670   0.307112  ...   1.175015   \n",
       "\n",
       "            4134740.0  7385511.0  7385515.0  7385547.0  7385552.0  7385611.0  \\\n",
       "GSM1914807   0.321372  -0.090157   0.617512  -0.645097  -0.851523  -1.100261   \n",
       "GSM1914808  -0.699605   1.423479   0.468561   0.082516  -0.001238  -1.229338   \n",
       "GSM1914809   0.214251  -0.027840   0.457613  -0.463707  -0.587461  -0.553096   \n",
       "GSM1914810  -0.812398   0.184977   0.633738   1.122524  -0.478908  -1.820808   \n",
       "GSM1914811   0.732782   0.310884  -1.773219  -0.204938   1.053665   1.811451   \n",
       "\n",
       "            7385641.0  7385683.0  7385696.0  \n",
       "GSM1914807  -0.583777   0.334289  -1.151476  \n",
       "GSM1914808   1.906317  -1.284613  -1.936364  \n",
       "GSM1914809   1.469179   0.052834  -1.200240  \n",
       "GSM1914810   0.285067   3.199836  -0.330872  \n",
       "GSM1914811   0.039464  -1.424448   0.255503  \n",
       "\n",
       "[5 rows x 22011 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process each matrix and filter based on index\n",
    "\n",
    "l_norm_X = pre_process_X(normalisedCounts)\n",
    "l_norm_X.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fourth-coaching",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(105, 22011)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_norm_X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "elder-japanese",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of dataset:   (105, 22011)\n",
      "Number of Missing values in dataset:   0\n",
      "The number of features with all zeros is:   0\n",
      "Check data and annotation indices match:   0\n"
     ]
    }
   ],
   "source": [
    "# get check output dimensions\n",
    "print('Shape of dataset:   '+str(l_norm_X.shape))\n",
    "print('Number of Missing values in dataset:   '+str(l_norm_X.isnull().sum().sum()))\n",
    "print('The number of features with all zeros is:   '+str(sum((l_norm_X == 0).all(axis=0))))\n",
    "print('Check data and annotation indices match:   '+str(sum(l_norm_X.index != labels_data.index)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "young-library",
   "metadata": {},
   "source": [
    "### Class Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "minute-teacher",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter labels data to cross ref\n",
    "labels_data\n",
    "\n",
    "# create labels vector\n",
    "labels_list = list(zip(labels_data.loc[:,'sampleID'], labels_data.loc[:,'condition']))\n",
    "labels = []\n",
    "for item in labels_list:\n",
    "    if item[1] == 'Sepsis':\n",
    "        labels.append(1)\n",
    "    else:\n",
    "        labels.append(0)\n",
    "labels = np.asarray(labels)\n",
    "labels_data['class'] = labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adjusted-madison",
   "metadata": {},
   "source": [
    "### Ensembl gene codes filter: affymetrix -> ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "willing-buyer",
   "metadata": {},
   "outputs": [],
   "source": [
    "## import raw gene code mappings from biomaRt\n",
    "gcMap = pd.read_csv('dataset_leo/affy_huex_1_0_st_v2_ensembl_mapping.csv').drop(['description', 'gene_biotype'], axis=1)\n",
    "#gcMap = gcMap.rename(columns={'affy_huex_1_0_st_v2': 'Probe_Id', 'ensembl_gene_id':'ensemblGeneID', 'external_gene_name':'geneName'})\n",
    "#gcMap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "adaptive-tactics",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>affy_huex_1_0_st_v2</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2467201</td>\n",
       "      <td>ENSG00000282052</td>\n",
       "      <td>RPS7P11</td>\n",
       "      <td>CHR_HSCHR17_2_CTG5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2467205</td>\n",
       "      <td>ENSG00000282052</td>\n",
       "      <td>RPS7P11</td>\n",
       "      <td>CHR_HSCHR17_2_CTG5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2467189</td>\n",
       "      <td>ENSG00000282052</td>\n",
       "      <td>RPS7P11</td>\n",
       "      <td>CHR_HSCHR17_2_CTG5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2467191</td>\n",
       "      <td>ENSG00000282052</td>\n",
       "      <td>RPS7P11</td>\n",
       "      <td>CHR_HSCHR17_2_CTG5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2523630</td>\n",
       "      <td>ENSG00000232733</td>\n",
       "      <td>RPL12P1</td>\n",
       "      <td>CHR_HSCHR6_MHC_SSTO_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>3807490</td>\n",
       "      <td>ENSG00000244716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>3807500</td>\n",
       "      <td>ENSG00000244716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>3807499</td>\n",
       "      <td>ENSG00000236768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>3807488</td>\n",
       "      <td>ENSG00000236768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>3809736</td>\n",
       "      <td>ENSG00000163482</td>\n",
       "      <td>STK36</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3318 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      affy_huex_1_0_st_v2  ensembl_gene_id external_gene_name  \\\n",
       "0                 2467201  ENSG00000282052            RPS7P11   \n",
       "1                 2467205  ENSG00000282052            RPS7P11   \n",
       "2                 2467189  ENSG00000282052            RPS7P11   \n",
       "3                 2467191  ENSG00000282052            RPS7P11   \n",
       "4                 2523630  ENSG00000232733            RPL12P1   \n",
       "...                   ...              ...                ...   \n",
       "3313              3807490  ENSG00000244716                NaN   \n",
       "3314              3807500  ENSG00000244716                NaN   \n",
       "3315              3807499  ENSG00000236768                NaN   \n",
       "3316              3807488  ENSG00000236768                NaN   \n",
       "3317              3809736  ENSG00000163482              STK36   \n",
       "\n",
       "               chromosome_name  \n",
       "0           CHR_HSCHR17_2_CTG5  \n",
       "1           CHR_HSCHR17_2_CTG5  \n",
       "2           CHR_HSCHR17_2_CTG5  \n",
       "3           CHR_HSCHR17_2_CTG5  \n",
       "4     CHR_HSCHR6_MHC_SSTO_CTG1  \n",
       "...                        ...  \n",
       "3313                         1  \n",
       "3314                         1  \n",
       "3315                         3  \n",
       "3316                         3  \n",
       "3317                         2  \n",
       "\n",
       "[3318 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gcMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "alike-roulette",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique probes in affymetrix dataset:                22011\n",
      "Number of lines of mappings in the mapping dataset:           3318\n",
      "Number of unique affymetrix probes in the dataset             1853\n",
      "Number of unique ensemble gene_Ids mapped to affymetrix probe:696\n",
      "Number of unique gene_names mapped to affymetrix probe:       524\n"
     ]
    }
   ],
   "source": [
    "# examine the gene code mappings file\n",
    "print('Number of unique probes in affymetrix dataset:                '+str(len(l_norm_X.columns.unique())))\n",
    "print('Number of lines of mappings in the mapping dataset:           '+str(len(gcMap['ensemblGeneID'])))\n",
    "print('Number of unique affymetrix probes in the dataset             '+str(len(gcMap['Probe_Id'].unique())))\n",
    "print('Number of unique ensemble gene_Ids mapped to affymetrix probe:'+str(len(gcMap['ensemblGeneID'].unique())))\n",
    "print('Number of unique gene_names mapped to affymetrix probe:       '+str(len(gcMap['geneName'].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "changed-scene",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of mappings in filtered list:                     2959\n",
      "Number of unique affymetrix probes mapped to ensembl id: 1853\n",
      "Number of unique ensemble gene_Ids mapped to affymetrix: 651\n",
      "Number of unique gene_names:                             524\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1853"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## filter out all probes that map to ensembel gene IDs not on the main chromosome (filter out Haplotyptic regions)\n",
    "## this doesn't reduce the number of genes in the annotation, but reduces duplicate ensembl IDs mapped to the same gene\n",
    "## source: https://www.researchgate.net/post/How-to-deal-with-multiple-ensemble-IDs-mapping-to-one-gene-symbol-in-a-RNA-Seq-dataset\n",
    "\n",
    "filtered_gcMap = gcMap[~gcMap['chromosome_name'].str.contains('CHR_')]\n",
    "\n",
    "\n",
    "print('Number of mappings in filtered list:                     '+str(len(filtered_gcMap)))\n",
    "print('Number of unique affymetrix probes mapped to ensembl id: '+str(len(filtered_gcMap['Probe_Id'].unique())))\n",
    "print('Number of unique ensemble gene_Ids mapped to affymetrix: '+str(len(filtered_gcMap['ensemblGeneID'].unique())))\n",
    "print('Number of unique gene_names:                             '+str(len(filtered_gcMap['geneName'].unique())))\n",
    "\n",
    "# get index of unique affymetrix probes to filter dataset by\n",
    "affy_index = filtered_gcMap['Probe_Id'].unique().tolist()\n",
    "len(affy_index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "urban-birmingham",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the filtered dataset:                     (1853, 105)\n",
      "Shape of the filtered and deduplicated dataset:    (105, 219)\n"
     ]
    }
   ],
   "source": [
    "# the main dataset needs to be filtered down from the 55k to 42k probes, excluding the 11k or so probes\n",
    "# that are not mapped to an ensemble ID\n",
    "\n",
    "l_filtered_X = l_norm_X.loc[:,affy_index].transpose()\n",
    "print('Shape of the filtered dataset:                     '+str(l_filtered_X.shape))\n",
    "\n",
    "# create dictionary of mapping between gene names to rename dataframes for analysis and visualisation\n",
    "# this sets the index of the dataframe to the probe_id, then pulls out the mappings - leaves out probes with no mapping\n",
    "mapping_dict_affy_en = pd.Series(filtered_gcMap.set_index('Probe_Id')['ensemblGeneID'].to_dict())\n",
    "\n",
    "# map the new column to the dataframe\n",
    "l_filtered_X['ensemblGeneID'] = mapping_dict_affy_en\n",
    "first_column = l_filtered_X.pop('ensemblGeneID')\n",
    "  \n",
    "## insert column using insert(position,column_name, first_column) function\n",
    "## take maximum expression values for duplicate ensembl_ID\n",
    "l_filtered_X.insert(0, 'ensemblGeneID', first_column)\n",
    "l_filtered_X = l_filtered_X.groupby('ensemblGeneID').max().sort_values('ensemblGeneID')\n",
    "l_filtered_X = l_filtered_X.transpose()\n",
    "\n",
    "print('Shape of the filtered and deduplicated dataset:    '+str(l_filtered_X.shape))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "standard-calendar",
   "metadata": {},
   "source": [
    "### Save to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "approximate-supply",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving pre-processed datasets\n",
    "\n",
    "l_filtered_X.to_csv('dataset_leo/ml_inputs/l_norm_X.csv')\n",
    "labels_data.to_csv('dataset_leo/ml_inputs/l_annotation.csv')\n",
    "np.savetxt('dataset_leo/ml_inputs/l_y.csv', labels, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "decreased-restriction",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
