{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "anonymous-tourism",
   "metadata": {},
   "source": [
    "# wong paediatrics dataset: pre-processing for ml input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cheap-garbage",
   "metadata": {},
   "source": [
    "### Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "hydraulic-washer",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import stats\n",
    "from collections import OrderedDict\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import warnings\n",
    "from joblib import dump, load\n",
    "from pickle import dump, load\n",
    "warnings.filterwarnings(action=\"ignore\", message=\"^internal gelsd\")\n",
    "#np.random.seed(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "behavioral-civilization",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/Ed/Documents/GitHub/sepsis_ml_omics_msc'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lesser-connectivity",
   "metadata": {},
   "source": [
    "### Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "handed-ebony",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Normalised counts data shape:  (54675, 247)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GSM1617492</th>\n",
       "      <th>GSM1617493</th>\n",
       "      <th>GSM1617494</th>\n",
       "      <th>GSM1617495</th>\n",
       "      <th>GSM1617496</th>\n",
       "      <th>GSM1617497</th>\n",
       "      <th>GSM1617498</th>\n",
       "      <th>GSM1617499</th>\n",
       "      <th>GSM1617500</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1617758</th>\n",
       "      <th>GSM1617759</th>\n",
       "      <th>GSM1617760</th>\n",
       "      <th>GSM1617761</th>\n",
       "      <th>GSM1617762</th>\n",
       "      <th>GSM1617763</th>\n",
       "      <th>GSM1617764</th>\n",
       "      <th>GSM1617765</th>\n",
       "      <th>GSM1617766</th>\n",
       "      <th>GSM1617767</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>6.480428</td>\n",
       "      <td>6.309763</td>\n",
       "      <td>6.178332</td>\n",
       "      <td>6.429684</td>\n",
       "      <td>6.704960</td>\n",
       "      <td>6.761714</td>\n",
       "      <td>6.590785</td>\n",
       "      <td>6.393995</td>\n",
       "      <td>6.170426</td>\n",
       "      <td>...</td>\n",
       "      <td>7.102116</td>\n",
       "      <td>7.055780</td>\n",
       "      <td>6.913683</td>\n",
       "      <td>6.489448</td>\n",
       "      <td>6.500215</td>\n",
       "      <td>6.429381</td>\n",
       "      <td>6.992181</td>\n",
       "      <td>6.392452</td>\n",
       "      <td>6.490935</td>\n",
       "      <td>7.173261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>5.957581</td>\n",
       "      <td>5.910509</td>\n",
       "      <td>6.429563</td>\n",
       "      <td>6.074588</td>\n",
       "      <td>5.733254</td>\n",
       "      <td>5.748208</td>\n",
       "      <td>6.249178</td>\n",
       "      <td>6.096216</td>\n",
       "      <td>5.609574</td>\n",
       "      <td>...</td>\n",
       "      <td>5.933110</td>\n",
       "      <td>6.206179</td>\n",
       "      <td>5.930328</td>\n",
       "      <td>5.185324</td>\n",
       "      <td>6.302903</td>\n",
       "      <td>6.670980</td>\n",
       "      <td>6.201271</td>\n",
       "      <td>5.341256</td>\n",
       "      <td>6.615973</td>\n",
       "      <td>5.692392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>8.767280</td>\n",
       "      <td>9.219351</td>\n",
       "      <td>9.609292</td>\n",
       "      <td>8.779422</td>\n",
       "      <td>7.482476</td>\n",
       "      <td>7.329360</td>\n",
       "      <td>8.402740</td>\n",
       "      <td>9.652488</td>\n",
       "      <td>8.365256</td>\n",
       "      <td>...</td>\n",
       "      <td>7.133309</td>\n",
       "      <td>8.227304</td>\n",
       "      <td>8.230635</td>\n",
       "      <td>7.256653</td>\n",
       "      <td>9.514109</td>\n",
       "      <td>9.240488</td>\n",
       "      <td>10.481245</td>\n",
       "      <td>8.877384</td>\n",
       "      <td>10.170809</td>\n",
       "      <td>8.534553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>8.685310</td>\n",
       "      <td>8.267696</td>\n",
       "      <td>7.746195</td>\n",
       "      <td>7.751903</td>\n",
       "      <td>8.007167</td>\n",
       "      <td>8.148816</td>\n",
       "      <td>8.346299</td>\n",
       "      <td>8.203552</td>\n",
       "      <td>8.146709</td>\n",
       "      <td>...</td>\n",
       "      <td>7.942987</td>\n",
       "      <td>8.239259</td>\n",
       "      <td>8.305233</td>\n",
       "      <td>9.005083</td>\n",
       "      <td>8.087161</td>\n",
       "      <td>7.800295</td>\n",
       "      <td>7.995118</td>\n",
       "      <td>8.518748</td>\n",
       "      <td>7.800516</td>\n",
       "      <td>8.712825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>2.949759</td>\n",
       "      <td>3.009085</td>\n",
       "      <td>2.660869</td>\n",
       "      <td>2.950493</td>\n",
       "      <td>2.843097</td>\n",
       "      <td>3.199670</td>\n",
       "      <td>2.891600</td>\n",
       "      <td>2.896296</td>\n",
       "      <td>2.872295</td>\n",
       "      <td>...</td>\n",
       "      <td>2.920644</td>\n",
       "      <td>3.024307</td>\n",
       "      <td>3.092109</td>\n",
       "      <td>3.293336</td>\n",
       "      <td>2.854140</td>\n",
       "      <td>2.898836</td>\n",
       "      <td>2.930668</td>\n",
       "      <td>3.019821</td>\n",
       "      <td>2.850628</td>\n",
       "      <td>2.988359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 247 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  GSM1617492  GSM1617493  GSM1617494  GSM1617495  GSM1617496  \\\n",
       "0  1007_s_at    6.480428    6.309763    6.178332    6.429684    6.704960   \n",
       "1    1053_at    5.957581    5.910509    6.429563    6.074588    5.733254   \n",
       "2     117_at    8.767280    9.219351    9.609292    8.779422    7.482476   \n",
       "3     121_at    8.685310    8.267696    7.746195    7.751903    8.007167   \n",
       "4  1255_g_at    2.949759    3.009085    2.660869    2.950493    2.843097   \n",
       "\n",
       "   GSM1617497  GSM1617498  GSM1617499  GSM1617500  ...  GSM1617758  \\\n",
       "0    6.761714    6.590785    6.393995    6.170426  ...    7.102116   \n",
       "1    5.748208    6.249178    6.096216    5.609574  ...    5.933110   \n",
       "2    7.329360    8.402740    9.652488    8.365256  ...    7.133309   \n",
       "3    8.148816    8.346299    8.203552    8.146709  ...    7.942987   \n",
       "4    3.199670    2.891600    2.896296    2.872295  ...    2.920644   \n",
       "\n",
       "   GSM1617759  GSM1617760  GSM1617761  GSM1617762  GSM1617763  GSM1617764  \\\n",
       "0    7.055780    6.913683    6.489448    6.500215    6.429381    6.992181   \n",
       "1    6.206179    5.930328    5.185324    6.302903    6.670980    6.201271   \n",
       "2    8.227304    8.230635    7.256653    9.514109    9.240488   10.481245   \n",
       "3    8.239259    8.305233    9.005083    8.087161    7.800295    7.995118   \n",
       "4    3.024307    3.092109    3.293336    2.854140    2.898836    2.930668   \n",
       "\n",
       "   GSM1617765  GSM1617766  GSM1617767  \n",
       "0    6.392452    6.490935    7.173261  \n",
       "1    5.341256    6.615973    5.692392  \n",
       "2    8.877384   10.170809    8.534553  \n",
       "3    8.518748    7.800516    8.712825  \n",
       "4    3.019821    2.850628    2.988359  \n",
       "\n",
       "[5 rows x 247 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import and check data\n",
    "\n",
    "# alternative X matrices\n",
    "normalisedCounts = pd.read_csv('dataset_wong/output/normalised_exp.csv') #, index_col=0)\n",
    "\n",
    "# annotation and class labels\n",
    "labels_data = pd.read_csv('dataset_wong/output/targets.csv', index_col=0)\n",
    "\n",
    "# check\n",
    "print('Normalised counts data shape:  '+str(normalisedCounts.shape))\n",
    "normalisedCounts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "raised-mustang",
   "metadata": {},
   "source": [
    "### Pre-process X matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "nutritional-racing",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pre-processing function for all the perth datasets types\n",
    "\n",
    "def pre_process_X(raw_data):\n",
    "    \n",
    "    ''' dataframe -> datafram\n",
    "    Input dataframe of normalised rnaseq data, transformed and z-scored\n",
    "    '''\n",
    "    # rename first column to ensembleGeneID and drop any duplicate rows based on this column\n",
    "    df = raw_data.rename({\"Unnamed: 0\": \"affy_probe_ID\"}, axis=1).drop_duplicates(subset=[\"affy_probe_ID\"], keep='first')\n",
    "    \n",
    "    # transpose the dataframe so genes are the columns\n",
    "    df = df.transpose()\n",
    "    \n",
    "    # set the column names to first row, i.e ensembleGeneID, in the transposed frame\n",
    "    df = df.rename(columns=df.iloc[0]).drop(df.index[0])\n",
    "    \n",
    "    # drop columns containing any NaN values\n",
    "    nan_columns = df.columns[df.isna().any()].tolist()\n",
    "    df = df.drop(nan_columns, axis=1)\n",
    "    \n",
    "    # z-score the data\n",
    "    scaler = StandardScaler()\n",
    "    df_scaled = scaler.fit_transform(df)\n",
    "    df = pd.DataFrame(df_scaled, columns=df.columns, index=df.index)\n",
    "    \n",
    "    return df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "seventh-appreciation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>1405_i_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1617492</th>\n",
       "      <td>0.204287</td>\n",
       "      <td>-0.646879</td>\n",
       "      <td>-0.185210</td>\n",
       "      <td>1.775490</td>\n",
       "      <td>-0.158048</td>\n",
       "      <td>-0.232908</td>\n",
       "      <td>-0.052714</td>\n",
       "      <td>1.228258</td>\n",
       "      <td>0.727841</td>\n",
       "      <td>0.097294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.237008</td>\n",
       "      <td>0.239686</td>\n",
       "      <td>-0.364037</td>\n",
       "      <td>-0.223004</td>\n",
       "      <td>-1.106239</td>\n",
       "      <td>-0.497183</td>\n",
       "      <td>1.715007</td>\n",
       "      <td>-0.785625</td>\n",
       "      <td>0.694494</td>\n",
       "      <td>-0.074895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1617493</th>\n",
       "      <td>-0.261806</td>\n",
       "      <td>-0.742750</td>\n",
       "      <td>0.286141</td>\n",
       "      <td>0.590625</td>\n",
       "      <td>0.279996</td>\n",
       "      <td>0.351596</td>\n",
       "      <td>-0.364469</td>\n",
       "      <td>-0.098497</td>\n",
       "      <td>1.154687</td>\n",
       "      <td>0.307041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.376046</td>\n",
       "      <td>0.412585</td>\n",
       "      <td>-0.712371</td>\n",
       "      <td>-0.135464</td>\n",
       "      <td>-1.203162</td>\n",
       "      <td>0.400214</td>\n",
       "      <td>-0.091309</td>\n",
       "      <td>-0.053911</td>\n",
       "      <td>-0.411598</td>\n",
       "      <td>-0.683423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1617494</th>\n",
       "      <td>-0.620747</td>\n",
       "      <td>0.314393</td>\n",
       "      <td>0.692711</td>\n",
       "      <td>-0.888989</td>\n",
       "      <td>-2.291128</td>\n",
       "      <td>0.792257</td>\n",
       "      <td>-1.311134</td>\n",
       "      <td>1.319395</td>\n",
       "      <td>1.314919</td>\n",
       "      <td>0.445230</td>\n",
       "      <td>...</td>\n",
       "      <td>0.574173</td>\n",
       "      <td>0.683378</td>\n",
       "      <td>0.381488</td>\n",
       "      <td>0.429228</td>\n",
       "      <td>-1.583988</td>\n",
       "      <td>-0.869088</td>\n",
       "      <td>-0.026533</td>\n",
       "      <td>-0.042388</td>\n",
       "      <td>-0.980302</td>\n",
       "      <td>0.781058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1617495</th>\n",
       "      <td>0.065702</td>\n",
       "      <td>-0.408575</td>\n",
       "      <td>-0.172549</td>\n",
       "      <td>-0.872795</td>\n",
       "      <td>-0.152624</td>\n",
       "      <td>0.793982</td>\n",
       "      <td>-1.166853</td>\n",
       "      <td>0.305920</td>\n",
       "      <td>1.145665</td>\n",
       "      <td>-0.269316</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.319668</td>\n",
       "      <td>-0.410791</td>\n",
       "      <td>-0.857264</td>\n",
       "      <td>-0.848270</td>\n",
       "      <td>-0.180219</td>\n",
       "      <td>-0.653681</td>\n",
       "      <td>-0.827999</td>\n",
       "      <td>0.392818</td>\n",
       "      <td>-0.536850</td>\n",
       "      <td>0.191827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1617496</th>\n",
       "      <td>0.817492</td>\n",
       "      <td>-1.103759</td>\n",
       "      <td>-1.524803</td>\n",
       "      <td>-0.148555</td>\n",
       "      <td>-0.945609</td>\n",
       "      <td>-0.340107</td>\n",
       "      <td>-0.419090</td>\n",
       "      <td>0.661588</td>\n",
       "      <td>0.228307</td>\n",
       "      <td>-0.060534</td>\n",
       "      <td>...</td>\n",
       "      <td>1.436831</td>\n",
       "      <td>1.801459</td>\n",
       "      <td>1.046046</td>\n",
       "      <td>1.234849</td>\n",
       "      <td>-0.132124</td>\n",
       "      <td>-0.566261</td>\n",
       "      <td>-1.193691</td>\n",
       "      <td>-1.172549</td>\n",
       "      <td>-0.344700</td>\n",
       "      <td>1.201937</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 54675 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            1007_s_at   1053_at    117_at    121_at  1255_g_at   1294_at  \\\n",
       "GSM1617492   0.204287 -0.646879 -0.185210  1.775490  -0.158048 -0.232908   \n",
       "GSM1617493  -0.261806 -0.742750  0.286141  0.590625   0.279996  0.351596   \n",
       "GSM1617494  -0.620747  0.314393  0.692711 -0.888989  -2.291128  0.792257   \n",
       "GSM1617495   0.065702 -0.408575 -0.172549 -0.872795  -0.152624  0.793982   \n",
       "GSM1617496   0.817492 -1.103759 -1.524803 -0.148555  -0.945609 -0.340107   \n",
       "\n",
       "             1316_at   1320_at  1405_i_at   1431_at  ...  \\\n",
       "GSM1617492 -0.052714  1.228258   0.727841  0.097294  ...   \n",
       "GSM1617493 -0.364469 -0.098497   1.154687  0.307041  ...   \n",
       "GSM1617494 -1.311134  1.319395   1.314919  0.445230  ...   \n",
       "GSM1617495 -1.166853  0.305920   1.145665 -0.269316  ...   \n",
       "GSM1617496 -0.419090  0.661588   0.228307 -0.060534  ...   \n",
       "\n",
       "            AFFX-r2-Ec-bioD-3_at  AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  \\\n",
       "GSM1617492              0.237008              0.239686            -0.364037   \n",
       "GSM1617493              0.376046              0.412585            -0.712371   \n",
       "GSM1617494              0.574173              0.683378             0.381488   \n",
       "GSM1617495             -0.319668             -0.410791            -0.857264   \n",
       "GSM1617496              1.436831              1.801459             1.046046   \n",
       "\n",
       "            AFFX-r2-P1-cre-5_at  AFFX-ThrX-3_at  AFFX-ThrX-5_at  \\\n",
       "GSM1617492            -0.223004       -1.106239       -0.497183   \n",
       "GSM1617493            -0.135464       -1.203162        0.400214   \n",
       "GSM1617494             0.429228       -1.583988       -0.869088   \n",
       "GSM1617495            -0.848270       -0.180219       -0.653681   \n",
       "GSM1617496             1.234849       -0.132124       -0.566261   \n",
       "\n",
       "            AFFX-ThrX-M_at  AFFX-TrpnX-3_at  AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "GSM1617492        1.715007        -0.785625         0.694494        -0.074895  \n",
       "GSM1617493       -0.091309        -0.053911        -0.411598        -0.683423  \n",
       "GSM1617494       -0.026533        -0.042388        -0.980302         0.781058  \n",
       "GSM1617495       -0.827999         0.392818        -0.536850         0.191827  \n",
       "GSM1617496       -1.193691        -1.172549        -0.344700         1.201937  \n",
       "\n",
       "[5 rows x 54675 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process each matrix and filter based on index\n",
    "\n",
    "w_norm_X = pre_process_X(normalisedCounts)\n",
    "w_norm_X.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "frank-blake",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(246, 54675)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "w_norm_X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "elder-japanese",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of dataset:   (246, 54675)\n",
      "Number of Missing values in dataset:   0\n",
      "The number of features with all zeros is:   0\n",
      "Check data and annotation indices match:   0\n"
     ]
    }
   ],
   "source": [
    "# get check output dimensions\n",
    "print('Shape of dataset:   '+str(w_norm_X.shape))\n",
    "print('Number of Missing values in dataset:   '+str(w_norm_X.isnull().sum().sum()))\n",
    "print('The number of features with all zeros is:   '+str(sum((w_norm_X == 0).all(axis=0))))\n",
    "print('Check data and annotation indices match:   '+str(sum(w_norm_X.index != labels_data.index)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "young-library",
   "metadata": {},
   "source": [
    "### Class Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "minute-teacher",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter labels data to cross ref\n",
    "labels_data\n",
    "\n",
    "# create labels vector\n",
    "labels_list = list(zip(labels_data.loc[:,'sampleID'], labels_data.loc[:,'condition']))\n",
    "labels = []\n",
    "for item in labels_list:\n",
    "    if item[1] == 'Control':\n",
    "        labels.append(0)\n",
    "    else:\n",
    "        labels.append(1)\n",
    "labels = np.asarray(labels)\n",
    "labels_data['class'] = labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "damaged-station",
   "metadata": {},
   "source": [
    "### Ensembl gene codes filter: affymetrix -> ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "sustainable-peninsula",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Probe_Id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1553551_s_at</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1553551_s_at</td>\n",
       "      <td>ENSG00000210100</td>\n",
       "      <td>MT-TI</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1553551_s_at</td>\n",
       "      <td>ENSG00000210112</td>\n",
       "      <td>MT-TM</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1553551_s_at</td>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1553538_s_at</td>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Probe_Id  ensembl_gene_id external_gene_name chromosome_name\n",
       "0  1553551_s_at  ENSG00000198888             MT-ND1              MT\n",
       "1  1553551_s_at  ENSG00000210100              MT-TI              MT\n",
       "2  1553551_s_at  ENSG00000210112              MT-TM              MT\n",
       "3  1553551_s_at  ENSG00000198763             MT-ND2              MT\n",
       "4  1553538_s_at  ENSG00000198804             MT-CO1              MT"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## import raw gene code mappings from biomaRt\n",
    "gcMap = pd.read_csv('dataset_wong/Affy_HG-U133_Plus_2_ensembl_mapping.csv').drop(['description', 'gene_biotype'], axis=1)\n",
    "gcMap = gcMap.rename(columns={'affy_hg_u133_plus_2': 'Probe_Id'})\n",
    "gcMap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "prime-species",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique probes in affymetrix dataset:                54675\n",
      "Number of lines of mappings in the mapping dataset:           52755\n",
      "Number of unique affymetrix probes in the dataset             42839\n",
      "Number of unique ensemble gene_Ids mapped to affymetrix probe:28215\n",
      "Number of unique gene_names mapped to affymetrix probe:       22084\n"
     ]
    }
   ],
   "source": [
    "# examine the gene code mappings file\n",
    "print('Number of unique probes in affymetrix dataset:                '+str(len(w_norm_X.columns.unique())))\n",
    "print('Number of lines of mappings in the mapping dataset:           '+str(len(gcMap['ensembl_gene_id'])))\n",
    "print('Number of unique affymetrix probes in the dataset             '+str(len(gcMap['Probe_Id'].unique())))\n",
    "print('Number of unique ensemble gene_Ids mapped to affymetrix probe:'+str(len(gcMap['ensembl_gene_id'].unique())))\n",
    "print('Number of unique gene_names mapped to affymetrix probe:       '+str(len(gcMap['external_gene_name'].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "alpha-wagon",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of mappings in filtered list:                     48006\n",
      "Number of unique affymetrix Probe_Ids:                   42806\n",
      "Number of unique ensemble gene_Ids mapped to affymetrix: 25358\n",
      "Number of unique gene_names:                             22066\n"
     ]
    }
   ],
   "source": [
    "## filter out all probes that map to ensembel gene IDs not on the main chromosome (filter out Haplotyptic regions)\n",
    "## this doesn't reduce the number of genes in the annotation, but reduces duplicate ensembl IDs mapped to the same gene\n",
    "## source: https://www.researchgate.net/post/How-to-deal-with-multiple-ensemble-IDs-mapping-to-one-gene-symbol-in-a-RNA-Seq-dataset\n",
    "\n",
    "filtered_gcMap = gcMap[~gcMap['chromosome_name'].str.contains('CHR_')]\n",
    "\n",
    "\n",
    "print('Number of mappings in filtered list:                     '+str(len(filtered_gcMap)))\n",
    "print('Number of unique affymetrix probes mapped to ensembl id: '+str(len(filtered_gcMap['Probe_Id'].unique())))\n",
    "print('Number of unique ensemble gene_Ids mapped to affymetrix: '+str(len(filtered_gcMap['ensembl_gene_id'].unique())))\n",
    "print('Number of unique gene_names:                             '+str(len(filtered_gcMap['external_gene_name'].unique())))\n",
    "\n",
    "# get index of unique affymetrix probes to filter dataset by\n",
    "affy_index = filtered_gcMap['Probe_Id'].unique().tolist()\n",
    "len(affy_index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "israeli-marriage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the filtered dataset:                     (42806, 246)\n",
      "Shape of the filtered and deduplicated dataset:    (246, 22966)\n"
     ]
    }
   ],
   "source": [
    "# the main dataset needs to be filtered down from the 55k to 42k probes, excluding the 11k or so probes\n",
    "# that are not mapped to an ensemble ID\n",
    "\n",
    "w_filtered_X = w_norm_X.loc[:,affy_index].transpose()\n",
    "print('Shape of the filtered dataset:                     '+str(w_filtered_X.shape))\n",
    "\n",
    "# create dictionary of mapping between gene names to rename dataframes for analysis and visualisation\n",
    "# this sets the index of the dataframe to the probe_id, then pulls out the mappings - leaves out probes with no mapping\n",
    "mapping_dict_affy_en = pd.Series(filtered_gcMap.set_index('Probe_Id')['ensembl_gene_id'].to_dict())\n",
    "\n",
    "# map the new column to the dataframe\n",
    "w_filtered_X['ensembl_ID'] = mapping_dict_affy_en\n",
    "first_column = w_filtered_X.pop('ensembl_ID')\n",
    "  \n",
    "## insert column using insert(position,column_name, first_column) function\n",
    "## take maximum expression values for duplicate ensembl_ID\n",
    "w_filtered_X.insert(0, 'ensembl_ID', first_column)\n",
    "w_filtered_X = w_filtered_X.groupby('ensembl_ID').max().sort_values('ensembl_ID')\n",
    "w_filtered_X = w_filtered_X.transpose()\n",
    "\n",
    "print('Shape of the filtered and deduplicated dataset:    '+str(w_filtered_X.shape))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "standard-calendar",
   "metadata": {},
   "source": [
    "### Save to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "approximate-supply",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving pre-processed datasets\n",
    "\n",
    "w_filtered_X.to_csv('dataset_wong/ml_inputs/w_norm_X.csv')\n",
    "labels_data.to_csv('dataset_wong/ml_inputs/w_annotation.csv')\n",
    "np.savetxt('dataset_wong/ml_inputs/w_y.csv', labels, delimiter=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vertical-battle",
   "metadata": {},
   "source": [
    "### sirs v sepsis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "ethical-sperm",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import and check data\n",
    "normalisedCounts_sirs = pd.read_csv('dataset_wong/output/normalised_exp_sirs.csv') #, index_col=0)\n",
    "labels_data_sirs = pd.read_csv('dataset_wong/output/targets_sirs.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "noble-checklist",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the filtered dataset:                     (42806, 229)\n",
      "Shape of the filtered and deduplicated dataset:    (229, 22966)\n"
     ]
    }
   ],
   "source": [
    "# preprocess X\n",
    "w_norm_X_sirs = pre_process_X(normalisedCounts_sirs)\n",
    "w_filtered_X_sirs = w_norm_X_sirs.loc[:,affy_index].transpose()\n",
    "print('Shape of the filtered dataset:                     '+str(w_filtered_X_sirs.shape))\n",
    "\n",
    "# map the new column to the dataframe\n",
    "w_filtered_X_sirs['ensembl_ID'] = mapping_dict_affy_en\n",
    "first_column = w_filtered_X_sirs.pop('ensembl_ID')\n",
    "w_filtered_X_sirs.insert(0, 'ensembl_ID', first_column)\n",
    "w_filtered_X_sirs = w_filtered_X_sirs.groupby('ensembl_ID').max().sort_values('ensembl_ID')\n",
    "w_filtered_X_sirs = w_filtered_X_sirs.transpose()\n",
    "print('Shape of the filtered and deduplicated dataset:    '+str(w_filtered_X_sirs.shape))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "electronic-explosion",
   "metadata": {},
   "outputs": [],
   "source": [
    "# process y\n",
    "labels_list_sirs = list(zip(labels_data_sirs.loc[:,'sampleID'], labels_data_sirs.loc[:,'condition']))\n",
    "labels_sirs = []\n",
    "for item in labels_list_sirs:\n",
    "    if item[1] == 'SIRS':\n",
    "        labels_sirs.append(0)\n",
    "    else:\n",
    "        labels_sirs.append(1)\n",
    "labels_sirs = np.asarray(labels_sirs)\n",
    "labels_data_sirs['class'] = labels_sirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "rocky-leave",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to csv\n",
    "w_filtered_X_sirs.to_csv('dataset_wong/ml_inputs/w_norm_sirs_X.csv')\n",
    "np.savetxt('dataset_wong/ml_inputs/w_sirs_y.csv', labels_sirs, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faced-helping",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
